In [2]:
import sys
import pandas as pd

sys.path.append("/var/www/python/Qingcheng/nighthawk/")
from nighthawk.util import sql_functions
from nighthawk.data.pipeline.common_functions.load import Load
from nighthawk.data.pipeline.common_functions.wind import Wind
from nighthawk.data.pipeline.common_functions.solar import Solar
from nighthawk.data.pipeline.common_functions.genoutage import GenOutage
from nighthawk.data.pipeline.var_handler.fuel_type import FuelType

START = "2026-04-01"
END   = "2026-04-01"
BAA_ZONES = ["E", "W"]

load     = Load("SPP")
wind     = Wind("SPP")
solar    = Solar("SPP")
genout   = GenOutage("SPP")
fueltype = FuelType("SPP")
print(f"Objects created for SPP  |  {START} → {END}")

Objects created for SPP  |  2026-04-01 → 2026-04-01


## 1. Total Load

In [2]:
df_total = load.get_total_load(START, END, var_spec=['f', 'a'])
print(df_total.shape)
df_total

(24, 4)


,dt,hr,spp_load_total_forecast_f,spp_load_total_actual_a
0,2026-04-01,1,32440.0,30110.0
1,2026-04-01,2,31585.0,31552.0
2,2026-04-01,3,31124.0,30860.0
3,2026-04-01,4,30883.0,30524.0
4,2026-04-01,5,31120.0,30709.0
5,2026-04-01,6,32023.0,31673.0
6,2026-04-01,7,33790.0,33342.0
7,2026-04-01,8,35016.0,34568.0
8,2026-04-01,9,35513.0,35212.0
9,2026-04-01,10,35851.0,35849.0


## 2. BAA Zonal Load (E and W)

In [3]:
df_baa = load.get_baa_zonal_load(START, END, baa_zones=BAA_ZONES, var_spec=['f', 'a'])
print(df_baa.shape)
df_baa

(48, 5)


,dt,hr,baa_zone,spp_baa_zonal_load_forecast_f,spp_baa_zonal_load_actual_a
0,2026-04-01,1,E,29713.0,27477.0
1,2026-04-01,1,W,2727.0,2633.0
2,2026-04-01,2,E,28966.0,29008.0
3,2026-04-01,2,W,2619.0,2544.0
4,2026-04-01,3,E,28563.0,28359.0
5,2026-04-01,3,W,2561.0,2501.0
6,2026-04-01,4,E,28340.0,28048.0
7,2026-04-01,4,W,2543.0,2476.0
8,2026-04-01,5,E,28571.0,28264.0
9,2026-04-01,5,W,2549.0,2445.0


## 3. Broadzonal Load

In [4]:
df_broadzonal = load.get_broadzonal_load(START, END, var_spec=['f', 'a'])
print(df_broadzonal.shape)
df_broadzonal

(72, 5)


,dt,hr,broadzone,spp_broadzonal_load_forecast_f,spp_broadzonal_load_actual_a
0,2026-04-01,1,NORTH,7650.74,7650.74
1,2026-04-01,1,SOUTH,21333.99,21333.99
2,2026-04-01,1,WEST,0.00,0.00
3,2026-04-01,2,NORTH,7579.01,7579.01
4,2026-04-01,2,SOUTH,20855.13,20855.13
...,...,...,...,...,...
67,2026-04-01,23,SOUTH,24017.62,NaN
68,2026-04-01,23,WEST,0.00,NaN
69,2026-04-01,24,NORTH,7796.01,NaN
70,2026-04-01,24,SOUTH,22593.36,NaN


## 4. Zonal Load

In [5]:
df_zonal = load.get_zonal_load(START, END, var_spec=['f', 'a'])
print(df_zonal.shape)
df_zonal

/var/www/python/Prod/nighthawk/nighthawk/data/pipeline/common_functions/load.py:1588: UserWarning: No zonal load forecast available for zone(s) ['PSCO', 'WACM', 'WAUW']; using actual load as forecast.
  warnings.warn(


(480, 5)


,dt,hr,zone,spp_zonal_load_forecast_f,spp_zonal_load_actual_a
0,2026-04-01,1,CSWS,4658.57,4658.57
1,2026-04-01,1,EDE,448.35,448.35
2,2026-04-01,1,GRDA,987.37,987.37
3,2026-04-01,1,INDN,79.79,79.79
4,2026-04-01,1,KACY,218.90,218.90
...,...,...,...,...,...
475,2026-04-01,24,WACM,NaN,NaN
476,2026-04-01,24,WAUE,3911.98,NaN
477,2026-04-01,24,WAUW,NaN,NaN
478,2026-04-01,24,WFEC,977.51,NaN


## 5. Reserve Zonal Load

In [6]:
df_res = load.get_res_zonal_load(START, END, var_spec=['f', 'a'])
print(df_res.shape)
df_res

(144, 5)


,dt,hr,ReserveZone,spp_res_zonal_load_forecast_f,spp_res_zonal_load_actual_a
0,2026-04-01,1,1,3660.03,3660.03
1,2026-04-01,1,2,657.00,657.00
2,2026-04-01,1,3,4096.22,4096.22
3,2026-04-01,1,4,16580.77,16580.77
4,2026-04-01,1,5,3990.71,3990.71
...,...,...,...,...,...
139,2026-04-01,24,2,680.42,NaN
140,2026-04-01,24,3,4173.73,NaN
141,2026-04-01,24,4,17739.21,NaN
142,2026-04-01,24,5,3911.98,NaN


## 6. Load Error

In [103]:
df_load_error = load.get_load_error(START, END, var_spec=['f', 'a'])
print(df_load_error.shape)
df_load_error

(24, 4)


,dt,hr,spp_load_error_f,spp_load_error_a
0,2026-04-01,1,-75.89660,-2330.0
1,2026-04-01,2,-145.97000,-33.0
2,2026-04-01,3,-116.74600,-264.0
3,2026-04-01,4,-80.73860,-359.0
4,2026-04-01,5,-117.05800,-411.0
5,2026-04-01,6,156.23500,-350.0
6,2026-04-01,7,228.36400,-448.0
7,2026-04-01,8,186.54900,-448.0
8,2026-04-01,9,26.12560,-301.0
9,2026-04-01,10,-41.49000,-2.0


## 7. BAA Zonal Load Error (E and W)

In [8]:
df_baa_error = load.get_baa_zonal_load_error(START, END, baa_zones=BAA_ZONES, var_spec=['f', 'a'])
print(df_baa_error.shape)
df_baa_error

(48, 5)


/var/www/python/Prod/nighthawk/nighthawk/data/pipeline/common_functions/load.py:2150: UserWarning: No model-predicted BAA-zonal load error forecast available for zone(s) ['W']; using actual load error as forecast.
  warnings.warn(


,dt,hr,baa_zone,spp_baa_zonal_load_error_a,spp_baa_zonal_load_error_f
0,2026-04-01,1,E,-2236.0,-158.31000
1,2026-04-01,1,W,-94.0,-94.00000
2,2026-04-01,2,E,42.0,-203.91600
3,2026-04-01,2,W,-75.0,-75.00000
4,2026-04-01,3,E,-204.0,-204.09600
5,2026-04-01,3,W,-60.0,-60.00000
6,2026-04-01,4,E,-292.0,-141.91500
7,2026-04-01,4,W,-67.0,-67.00000
8,2026-04-01,5,E,-307.0,-179.26300
9,2026-04-01,5,W,-104.0,-104.00000


---
## WIND

### 8. Total Wind

In [3]:
df_wind_total = wind.get_total_wind(START, END, var_spec=['f', 'a'])
print(df_wind_total.shape)
df_wind_total

(24, 4)


,dt,hr,spp_wind_total_forecast_f,spp_wind_total_actual_a
0,2026-04-01,1,19661.68,17018.20
1,2026-04-01,2,18941.02,16073.57
2,2026-04-01,3,18526.22,15514.78
3,2026-04-01,4,18392.06,14693.56
4,2026-04-01,5,18447.30,13884.95
5,2026-04-01,6,18251.26,13138.94
6,2026-04-01,7,17755.26,12729.71
7,2026-04-01,8,17281.17,11809.35
8,2026-04-01,9,16633.53,9874.16
9,2026-04-01,10,16403.86,8009.48


### 9. BAA Zonal Wind (E and W)

In [4]:
df_wind_baa = wind.get_baa_zonal_wind(START, END, baa_zones=BAA_ZONES, var_spec=['f', 'a'])
print(df_wind_baa.shape)
df_wind_baa

(48, 5)


,dt,hr,baa_zone,spp_baa_zonal_wind_forecast_f,spp_baa_zonal_wind_actual_a
0,2026-04-01,1,E,18820.52,16197.00
1,2026-04-01,1,W,841.16,821.20
2,2026-04-01,2,E,18101.49,15325.80
3,2026-04-01,2,W,839.53,747.77
4,2026-04-01,3,E,17711.71,14835.12
5,2026-04-01,3,W,814.51,679.66
6,2026-04-01,4,E,17604.79,14035.19
7,2026-04-01,4,W,787.27,658.37
8,2026-04-01,5,E,17679.20,13225.89
9,2026-04-01,5,W,768.10,659.06


### 10. Reserve Zonal Wind

In [18]:
df_wind_res = wind.get_res_zonal_wind(START, END, var_spec=['f', 'a'])
display(df_wind_res.head(10))
df_wind_res

,dt,hr,ReserveZone,spp_res_zonal_wind_forecast_f,spp_res_zonal_wind_actual_a
0,2026-04-01,1,1,1907.53,1641.80
1,2026-04-01,1,2,3762.81,3894.99
2,2026-04-01,1,3,1125.62,921.67
3,2026-04-01,1,4,9296.98,7498.03
4,2026-04-01,1,5,2727.58,2160.04
5,2026-04-01,1,21,841.16,819.10
6,2026-04-01,2,1,1809.51,1570.33
7,2026-04-01,2,2,3436.82,3491.59
8,2026-04-01,2,3,995.66,591.17
9,2026-04-01,2,4,9127.73,7483.80


,dt,hr,ReserveZone,spp_res_zonal_wind_forecast_f,spp_res_zonal_wind_actual_a
0,2026-04-01,1,1,1907.53,1641.80
1,2026-04-01,1,2,3762.81,3894.99
2,2026-04-01,1,3,1125.62,921.67
3,2026-04-01,1,4,9296.98,7498.03
4,2026-04-01,1,5,2727.58,2160.04
...,...,...,...,...,...
139,2026-04-01,24,2,5785.92,4455.42
140,2026-04-01,24,3,2009.54,1653.04
141,2026-04-01,24,4,13878.44,9536.89
142,2026-04-01,24,5,2635.36,2611.94


### 11. Wind Error

In [6]:
df_wind_error = wind.get_wind_error(START, END, var_spec=['f', 'a'])
print(df_wind_error.shape)
df_wind_error

(24, 4)


,dt,hr,spp_wind_error_f,spp_wind_error_a
0,2026-04-01,1,639.7140,2453.836674
1,2026-04-01,2,639.7140,2695.469547
2,2026-04-01,3,639.7140,2991.373332
3,2026-04-01,4,634.2130,3698.455000
4,2026-04-01,5,616.3100,4562.114167
5,2026-04-01,6,600.6570,5110.630000
6,2026-04-01,7,600.6570,5024.201667
7,2026-04-01,8,606.1580,5448.957501
8,2026-04-01,9,606.1580,6757.455000
9,2026-04-01,10,600.7200,8370.631667


### 12. BAA Zonal Wind Error (E and W)

In [7]:
df_wind_baa_error = wind.get_baa_zonal_wind_error(START, END, baa_zones=BAA_ZONES, var_spec=['f', 'a'])
print(df_wind_baa_error.shape)
df_wind_baa_error

(48, 5)


/var/www/python/Prod/nighthawk/nighthawk/data/pipeline/common_functions/wind.py:1593: UserWarning: No model-predicted BAA-zonal wind error forecast available for zone(s) ['W']; using actual wind error as forecast.
  warnings.warn(


,dt,hr,baa_zone,spp_baa_zonal_wind_error_a,spp_baa_zonal_wind_error_f
0,2026-04-01,1,E,2433.876674,639.714000
1,2026-04-01,1,W,19.960000,19.960000
2,2026-04-01,2,E,2614.815000,639.714000
3,2026-04-01,2,W,80.654546,80.654546
4,2026-04-01,3,E,2858.216667,639.714000
5,2026-04-01,3,W,133.156667,133.156667
6,2026-04-01,4,E,3569.600000,634.213000
7,2026-04-01,4,W,128.855000,128.855000
8,2026-04-01,5,E,4453.074167,616.310000
9,2026-04-01,5,W,109.040000,109.040000


### 13. VER Curtailment

In [8]:
df_ver = wind.get_ver_curtailment(START, END, var_spec=['f', 'a'])
print(df_ver.shape)
df_ver

(24, 4)


,dt,hr,spp_ver_curtailments_forecast_f,spp_ver_curtailments_actual_a
0,2026-04-01,1,1811.040,189.643326
1,2026-04-01,2,1759.060,171.980453
2,2026-04-01,3,1525.740,20.066668
3,2026-04-01,4,1517.030,0.045000
4,2026-04-01,5,1510.380,0.235833
5,2026-04-01,6,1184.470,1.690000
6,2026-04-01,7,1122.280,1.348333
7,2026-04-01,8,1101.870,22.862499
8,2026-04-01,9,647.363,1.915000
9,2026-04-01,10,532.968,23.748333


### 14. BAA Zonal VER Curtailment (E and W)

In [26]:
df_ver_baa = wind.get_baa_zonal_ver_curtailment('2026-04-03', '2026-04-03', baa_zones=BAA_ZONES, var_spec=['f', 'a'])
print(df_ver_baa.shape)
df_ver_baa

(24, 5)


/var/www/python/Prod/nighthawk/nighthawk/data/pipeline/common_functions/wind.py:1799: UserWarning: No BAA-zonal VER curtailment forecast available for zone(s) ['W']; using actual curtailment as forecast.
  warnings.warn(


,dt,hr,baa_zone,spp_baa_zonal_ver_curtailments_forecast_f,spp_baa_zonal_ver_curtailments_actual_a
0,2026-04-03,1,E,6153.24,NaN
1,2026-04-03,2,E,6430.88,NaN
2,2026-04-03,3,E,6702.76,NaN
3,2026-04-03,4,E,6808.07,NaN
4,2026-04-03,5,E,6626.88,NaN
5,2026-04-03,6,E,5765.97,NaN
6,2026-04-03,7,E,5497.78,NaN
7,2026-04-03,8,E,5065.83,NaN
8,2026-04-03,9,E,4994.91,NaN
9,2026-04-03,10,E,3866.38,NaN


---
## SOLAR

### 15. Total Solar

In [10]:
df_solar_total = solar.get_total_solar(START, END, var_spec=['f', 'a'])
print(df_solar_total.shape)
df_solar_total

(24, 4)


,dt,hr,spp_solar_total_forecast_f,spp_solar_total_actual_a
0,2026-04-01,1,0.00,0.60
1,2026-04-01,2,0.00,0.60
2,2026-04-01,3,0.00,0.60
3,2026-04-01,4,0.00,0.60
4,2026-04-01,5,0.00,0.60
5,2026-04-01,6,0.00,0.60
6,2026-04-01,7,0.05,0.60
7,2026-04-01,8,53.81,22.74
8,2026-04-01,9,422.77,363.38
9,2026-04-01,10,827.84,744.52


### 16. BAA Zonal Solar (E and W)

In [11]:
df_solar_baa = solar.get_baa_zonal_solar(START, END, baa_zones=BAA_ZONES, var_spec=['f', 'a'])
print(df_solar_baa.shape)
df_solar_baa

(48, 5)


,dt,hr,baa_zone,spp_baa_zonal_solar_forecast_f,spp_baa_zonal_solar_actual_a
0,2026-04-01,1,E,0.60,0.60
1,2026-04-01,1,W,0.00,0.00
2,2026-04-01,2,E,0.60,0.60
3,2026-04-01,2,W,0.00,0.00
4,2026-04-01,3,E,0.60,0.60
5,2026-04-01,3,W,0.00,0.00
6,2026-04-01,4,E,0.60,0.60
7,2026-04-01,4,W,0.00,0.00
8,2026-04-01,5,E,0.60,0.60
9,2026-04-01,5,W,0.00,0.00


### 17. Reserve Zonal Solar

In [25]:
df_solar_res = solar.get_res_zonal_solar(START, END, var_spec=['f', 'a'])
print(df_solar_res.shape)
df_solar_res.sort_values(['dt','hr','ReserveZone'])[:10]

(144, 5)


,dt,hr,ReserveZone,spp_res_zonal_solar_forecast_f,spp_res_zonal_solar_actual_a
0,2026-04-01,1,1,0.0,0.00
24,2026-04-01,1,2,0.0,0.00
48,2026-04-01,1,3,0.0,-0.03
72,2026-04-01,1,4,0.0,0.60
96,2026-04-01,1,5,0.0,0.00
120,2026-04-01,1,21,0.0,0.00
1,2026-04-01,2,1,0.0,0.00
25,2026-04-01,2,2,0.0,0.00
49,2026-04-01,2,3,0.0,0.05
73,2026-04-01,2,4,0.0,0.60


---
## GEN OUTAGE

### 18. Gen Outage — Total SPP

In [13]:
df_genout_total = genout.get_genoutage_by_level(START, END, area_list=['SPP'], var_spec=['f', 'a'])
print(df_genout_total.shape)
df_genout_total

(24, 5)


,dt,hr,baa_zone,spp_genoutage_forecast_f,spp_genoutage_actual_a
0,2026-04-01,1,SPP,25715.9,25197.9
1,2026-04-01,2,SPP,25728.3,25198.1
2,2026-04-01,3,SPP,25878.9,25348.7
3,2026-04-01,4,SPP,25878.9,25348.7
4,2026-04-01,5,SPP,25712.9,25182.7
5,2026-04-01,6,SPP,25612.9,25182.7
6,2026-04-01,7,SPP,25897.1,25171.9
7,2026-04-01,8,SPP,25907.1,25179.9
8,2026-04-01,9,SPP,25907.1,25180.5
9,2026-04-01,10,SPP,25907.7,25183.5


### 19. Gen Outage — BAA Zonal (E and W)

In [1]:
import sys
import pandas as pd

sys.path.append("/var/www/python/Qingcheng/nighthawk/")
from nighthawk.util import sql_functions
from nighthawk.data.pipeline.common_functions.load import Load
from nighthawk.data.pipeline.common_functions.wind import Wind
from nighthawk.data.pipeline.common_functions.solar import Solar
from nighthawk.data.pipeline.common_functions.genoutage import GenOutage
from nighthawk.data.pipeline.var_handler.fuel_type import FuelType

START = "2026-04-01"
END   = "2026-04-01"
BAA_ZONES = ["E", "W"]

load     = Load("SPP")
wind     = Wind("SPP")
solar    = Solar("SPP")
genout   = GenOutage("SPP")
fueltype = FuelType("SPP")
print(f"Objects created for SPP  |  {START} → {END}")

df_genout_baa = genout.get_genoutage_by_level('2026-04-07', '2026-04-07', area_list=['E', 'W'], var_spec=['f', 'a'])
print(df_genout_baa.shape)
display(df_genout_baa)

df_outage_fuel = fueltype.get_genoutage_fuel_type('2026-04-07', '2026-04-07', var_spec=['f', 'a'])
print(df_outage_fuel.shape)

# df_outage_fuel['total_f'] = df_outage_fuel.filter(regex='_f$').sum(axis=1)
# df_outage_fuel['total_a'] = df_outage_fuel.filter(regex='_a$').sum(axis=1)
# df_outage_fuel

fueltype.get_genoutage_fuel_type('2026-04-07', '2026-04-07', var_spec=['f', 'a'])

Objects created for SPP  |  2026-04-01 → 2026-04-01
(48, 5)


,dt,hr,baa_zone,spp_genoutage_forecast_f,spp_genoutage_actual_a
0,2026-04-07,1,E,26588.4,27190.4
1,2026-04-07,1,W,1739.1,1858.0
2,2026-04-07,2,E,26790.4,27392.4
3,2026-04-07,2,W,1739.1,1858.0
4,2026-04-07,3,E,26790.4,27392.4
5,2026-04-07,3,W,1858.0,1858.0
6,2026-04-07,4,E,26790.4,27392.4
7,2026-04-07,4,W,1858.0,1858.0
8,2026-04-07,5,E,27392.4,26613.2
9,2026-04-07,5,W,1858.0,1939.6


,spp_fuelmixoutage_coal_forecast_f,spp_fuelmixoutage_dieselfueloil_forecast_f,spp_fuelmixoutage_hydro_forecast_f,spp_fuelmixoutage_naturalgas_forecast_f,spp_fuelmixoutage_solar_forecast_f,spp_fuelmixoutage_wind_forecast_f,dt,hr,spp_fuelmixoutage_coal_actual_a,spp_fuelmixoutage_dieselfueloil_actual_a,spp_fuelmixoutage_hydro_actual_a,spp_fuelmixoutage_naturalgas_actual_a,spp_fuelmixoutage_solar_actual_a,spp_fuelmixoutage_wind_actual_a
0,9532.8,75.8,1620.3,14487.3,820.5,263.0,2026-04-05,19,8995.8,75.8,1620.3,13875.3,820.5,263.0
1,9532.8,75.8,1620.3,14487.3,820.5,263.0,2026-04-05,20,8995.8,75.8,1620.3,13875.3,820.5,263.0
2,9532.8,75.8,1620.3,14487.3,820.5,263.0,2026-04-05,21,8995.8,75.8,1620.3,13875.3,820.5,263.0
3,9532.8,75.8,1620.3,14487.3,970.5,263.0,2026-04-05,22,8995.8,75.8,1620.3,13875.3,970.5,263.0
4,9532.8,75.8,1620.3,14487.3,970.5,263.0,2026-04-05,23,8995.8,75.8,1620.3,13875.3,970.5,263.0
5,9532.8,75.8,1620.3,14487.3,970.5,263.0,2026-04-05,24,8995.8,75.8,1620.3,13875.3,970.5,263.0
6,9685.8,75.8,1620.3,14934.2,970.5,263.0,2026-04-06,1,9774.8,75.8,1620.3,14253.7,970.5,263.0
7,9823.8,75.8,1620.3,14934.2,970.5,263.0,2026-04-06,2,9986.8,75.8,1620.3,14253.7,970.5,263.0
8,9942.7,75.8,1685.3,14934.2,970.5,263.0,2026-04-06,3,10105.7,75.8,1685.3,14253.7,970.5,263.0
9,9942.7,75.8,1698.1,14934.2,970.5,263.0,2026-04-06,4,10105.7,75.8,1698.1,14253.7,970.5,263.0


,dt,hr,spp_fuelmixoutage_coal_forecast_f,spp_fuelmixoutage_dieselfueloil_forecast_f,spp_fuelmixoutage_hydro_forecast_f,spp_fuelmixoutage_naturalgas_forecast_f,spp_fuelmixoutage_solar_forecast_f,spp_fuelmixoutage_wind_forecast_f,spp_fuelmixoutage_coal_actual_a,spp_fuelmixoutage_dieselfueloil_actual_a,spp_fuelmixoutage_hydro_actual_a,spp_fuelmixoutage_naturalgas_actual_a,spp_fuelmixoutage_solar_actual_a,spp_fuelmixoutage_wind_actual_a
0,2026-04-07,1,9912.8,119.8,1656.1,14793.8,1247.5,283.0,10736.7,119.8,1656.1,14688.8,1247.5,283.0
1,2026-04-07,2,9912.8,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
2,2026-04-07,3,10031.7,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
3,2026-04-07,4,10031.7,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
4,2026-04-07,5,10736.7,119.8,1656.1,14890.8,1247.5,283.0,9990.8,119.8,1656.1,14925.4,1255.0,283.0
5,2026-04-07,6,10736.7,119.8,1656.1,15031.8,1247.5,283.0,10265.8,119.8,1656.1,15066.4,1255.0,283.0
6,2026-04-07,7,10736.7,119.8,1676.1,15031.8,1247.5,283.0,10265.8,119.8,1676.1,15066.4,1255.0,283.0
7,2026-04-07,8,10736.7,119.8,1676.1,15377.8,1097.5,283.0,10265.8,119.8,1676.1,15870.4,1105.0,283.0
8,2026-04-07,9,10736.7,119.8,1676.1,15377.8,1097.5,878.0,10265.8,119.8,1688.1,15870.4,1105.0,878.0
9,2026-04-07,10,10736.7,119.8,1676.1,15377.8,1097.5,878.0,10265.8,119.8,1688.1,15870.4,1105.0,878.0


,dt,hr,spp_fuelmixoutage_coal_forecast_f,spp_fuelmixoutage_dieselfueloil_forecast_f,spp_fuelmixoutage_hydro_forecast_f,spp_fuelmixoutage_naturalgas_forecast_f,spp_fuelmixoutage_solar_forecast_f,spp_fuelmixoutage_wind_forecast_f,spp_fuelmixoutage_coal_actual_a,spp_fuelmixoutage_dieselfueloil_actual_a,spp_fuelmixoutage_hydro_actual_a,spp_fuelmixoutage_naturalgas_actual_a,spp_fuelmixoutage_solar_actual_a,spp_fuelmixoutage_wind_actual_a
0,2026-04-07,1,9912.8,119.8,1656.1,14793.8,1247.5,283.0,10736.7,119.8,1656.1,14688.8,1247.5,283.0
1,2026-04-07,2,9912.8,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
2,2026-04-07,3,10031.7,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
3,2026-04-07,4,10031.7,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
4,2026-04-07,5,10736.7,119.8,1656.1,14890.8,1247.5,283.0,9990.8,119.8,1656.1,14925.4,1255.0,283.0
5,2026-04-07,6,10736.7,119.8,1656.1,15031.8,1247.5,283.0,10265.8,119.8,1656.1,15066.4,1255.0,283.0
6,2026-04-07,7,10736.7,119.8,1676.1,15031.8,1247.5,283.0,10265.8,119.8,1676.1,15066.4,1255.0,283.0
7,2026-04-07,8,10736.7,119.8,1676.1,15377.8,1097.5,283.0,10265.8,119.8,1676.1,15870.4,1105.0,283.0
8,2026-04-07,9,10736.7,119.8,1676.1,15377.8,1097.5,878.0,10265.8,119.8,1688.1,15870.4,1105.0,878.0
9,2026-04-07,10,10736.7,119.8,1676.1,15377.8,1097.5,878.0,10265.8,119.8,1688.1,15870.4,1105.0,878.0


(24, 14)


,spp_fuelmixoutage_coal_forecast_f,spp_fuelmixoutage_dieselfueloil_forecast_f,spp_fuelmixoutage_hydro_forecast_f,spp_fuelmixoutage_naturalgas_forecast_f,spp_fuelmixoutage_solar_forecast_f,spp_fuelmixoutage_wind_forecast_f,dt,hr,spp_fuelmixoutage_coal_actual_a,spp_fuelmixoutage_dieselfueloil_actual_a,spp_fuelmixoutage_hydro_actual_a,spp_fuelmixoutage_naturalgas_actual_a,spp_fuelmixoutage_solar_actual_a,spp_fuelmixoutage_wind_actual_a
0,9532.8,75.8,1620.3,14487.3,820.5,263.0,2026-04-05,19,8995.8,75.8,1620.3,13875.3,820.5,263.0
1,9532.8,75.8,1620.3,14487.3,820.5,263.0,2026-04-05,20,8995.8,75.8,1620.3,13875.3,820.5,263.0
2,9532.8,75.8,1620.3,14487.3,820.5,263.0,2026-04-05,21,8995.8,75.8,1620.3,13875.3,820.5,263.0
3,9532.8,75.8,1620.3,14487.3,970.5,263.0,2026-04-05,22,8995.8,75.8,1620.3,13875.3,970.5,263.0
4,9532.8,75.8,1620.3,14487.3,970.5,263.0,2026-04-05,23,8995.8,75.8,1620.3,13875.3,970.5,263.0
5,9532.8,75.8,1620.3,14487.3,970.5,263.0,2026-04-05,24,8995.8,75.8,1620.3,13875.3,970.5,263.0
6,9685.8,75.8,1620.3,14934.2,970.5,263.0,2026-04-06,1,9774.8,75.8,1620.3,14253.7,970.5,263.0
7,9823.8,75.8,1620.3,14934.2,970.5,263.0,2026-04-06,2,9986.8,75.8,1620.3,14253.7,970.5,263.0
8,9942.7,75.8,1685.3,14934.2,970.5,263.0,2026-04-06,3,10105.7,75.8,1685.3,14253.7,970.5,263.0
9,9942.7,75.8,1698.1,14934.2,970.5,263.0,2026-04-06,4,10105.7,75.8,1698.1,14253.7,970.5,263.0


,dt,hr,spp_fuelmixoutage_coal_forecast_f,spp_fuelmixoutage_dieselfueloil_forecast_f,spp_fuelmixoutage_hydro_forecast_f,spp_fuelmixoutage_naturalgas_forecast_f,spp_fuelmixoutage_solar_forecast_f,spp_fuelmixoutage_wind_forecast_f,spp_fuelmixoutage_coal_actual_a,spp_fuelmixoutage_dieselfueloil_actual_a,spp_fuelmixoutage_hydro_actual_a,spp_fuelmixoutage_naturalgas_actual_a,spp_fuelmixoutage_solar_actual_a,spp_fuelmixoutage_wind_actual_a
0,2026-04-07,1,9912.8,119.8,1656.1,14793.8,1247.5,283.0,10736.7,119.8,1656.1,14688.8,1247.5,283.0
1,2026-04-07,2,9912.8,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
2,2026-04-07,3,10031.7,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
3,2026-04-07,4,10031.7,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
4,2026-04-07,5,10736.7,119.8,1656.1,14890.8,1247.5,283.0,9990.8,119.8,1656.1,14925.4,1255.0,283.0
5,2026-04-07,6,10736.7,119.8,1656.1,15031.8,1247.5,283.0,10265.8,119.8,1656.1,15066.4,1255.0,283.0
6,2026-04-07,7,10736.7,119.8,1676.1,15031.8,1247.5,283.0,10265.8,119.8,1676.1,15066.4,1255.0,283.0
7,2026-04-07,8,10736.7,119.8,1676.1,15377.8,1097.5,283.0,10265.8,119.8,1676.1,15870.4,1105.0,283.0
8,2026-04-07,9,10736.7,119.8,1676.1,15377.8,1097.5,878.0,10265.8,119.8,1688.1,15870.4,1105.0,878.0
9,2026-04-07,10,10736.7,119.8,1676.1,15377.8,1097.5,878.0,10265.8,119.8,1688.1,15870.4,1105.0,878.0


,dt,hr,spp_fuelmixoutage_coal_forecast_f,spp_fuelmixoutage_dieselfueloil_forecast_f,spp_fuelmixoutage_hydro_forecast_f,spp_fuelmixoutage_naturalgas_forecast_f,spp_fuelmixoutage_solar_forecast_f,spp_fuelmixoutage_wind_forecast_f,spp_fuelmixoutage_coal_actual_a,spp_fuelmixoutage_dieselfueloil_actual_a,spp_fuelmixoutage_hydro_actual_a,spp_fuelmixoutage_naturalgas_actual_a,spp_fuelmixoutage_solar_actual_a,spp_fuelmixoutage_wind_actual_a
0,2026-04-07,1,9912.8,119.8,1656.1,14793.8,1247.5,283.0,10736.7,119.8,1656.1,14688.8,1247.5,283.0
1,2026-04-07,2,9912.8,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
2,2026-04-07,3,10031.7,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
3,2026-04-07,4,10031.7,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
4,2026-04-07,5,10736.7,119.8,1656.1,14890.8,1247.5,283.0,9990.8,119.8,1656.1,14925.4,1255.0,283.0
5,2026-04-07,6,10736.7,119.8,1656.1,15031.8,1247.5,283.0,10265.8,119.8,1656.1,15066.4,1255.0,283.0
6,2026-04-07,7,10736.7,119.8,1676.1,15031.8,1247.5,283.0,10265.8,119.8,1676.1,15066.4,1255.0,283.0
7,2026-04-07,8,10736.7,119.8,1676.1,15377.8,1097.5,283.0,10265.8,119.8,1676.1,15870.4,1105.0,283.0
8,2026-04-07,9,10736.7,119.8,1676.1,15377.8,1097.5,878.0,10265.8,119.8,1688.1,15870.4,1105.0,878.0
9,2026-04-07,10,10736.7,119.8,1676.1,15377.8,1097.5,878.0,10265.8,119.8,1688.1,15870.4,1105.0,878.0


,dt,hr,spp_fuelmixoutage_coal_forecast_f,spp_fuelmixoutage_dieselfueloil_forecast_f,spp_fuelmixoutage_hydro_forecast_f,spp_fuelmixoutage_naturalgas_forecast_f,spp_fuelmixoutage_solar_forecast_f,spp_fuelmixoutage_wind_forecast_f,spp_fuelmixoutage_coal_actual_a,spp_fuelmixoutage_dieselfueloil_actual_a,spp_fuelmixoutage_hydro_actual_a,spp_fuelmixoutage_naturalgas_actual_a,spp_fuelmixoutage_solar_actual_a,spp_fuelmixoutage_wind_actual_a
0,2026-04-07,1,9912.8,119.8,1656.1,14793.8,1247.5,283.0,10736.7,119.8,1656.1,14688.8,1247.5,283.0
1,2026-04-07,2,9912.8,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
2,2026-04-07,3,10031.7,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
3,2026-04-07,4,10031.7,119.8,1656.1,14995.8,1247.5,283.0,10736.7,119.8,1656.1,14890.8,1247.5,283.0
4,2026-04-07,5,10736.7,119.8,1656.1,14890.8,1247.5,283.0,9990.8,119.8,1656.1,14925.4,1255.0,283.0
5,2026-04-07,6,10736.7,119.8,1656.1,15031.8,1247.5,283.0,10265.8,119.8,1656.1,15066.4,1255.0,283.0
6,2026-04-07,7,10736.7,119.8,1676.1,15031.8,1247.5,283.0,10265.8,119.8,1676.1,15066.4,1255.0,283.0
7,2026-04-07,8,10736.7,119.8,1676.1,15377.8,1097.5,283.0,10265.8,119.8,1676.1,15870.4,1105.0,283.0
8,2026-04-07,9,10736.7,119.8,1676.1,15377.8,1097.5,878.0,10265.8,119.8,1688.1,15870.4,1105.0,878.0
9,2026-04-07,10,10736.7,119.8,1676.1,15377.8,1097.5,878.0,10265.8,119.8,1688.1,15870.4,1105.0,878.0


---
## GEN MIX (FUEL TYPE)

### 20. All 5 GenMix Nighthawk Functions — 2026-04-01

| # | Function | Table | Fuel types | Zones | a/f |
|---|---|---|---|---|---|
| 1 | `get_gen_fuel_type` | `GenMixHourly` | Hydro, Nuclear, NaturalGas | SPP total | both |
| 2 | `get_gen_fuel_actual` | `GenMixHourly` | all 10 | SPP total | actual only |
| 3 | `get_baa_zonal_gen_fuel_type` | `BAAZonalGenMixHourly` | Hydro, Nuclear, NaturalGas | E + W | both |
| 4 | `get_baa_zonal_gen_fuel_actual` | `BAAZonalGenMixHourly` | all 10 | E + W | actual only |
| 5 | `get_spp_gas_output` *(deprecated)* | `GenMixHourly` | NaturalGas only | SPP total | both |

In [1]:
import sys
sys.path.append("/var/www/python/Qingcheng/nighthawk/")
from nighthawk.data.pipeline.var_handler.fuel_type import FuelType
from nighthawk.data.pipeline.var_handler.spp_specific.gas_price_effect import get_spp_gas_output

ft = FuelType('SPP')
D = '2026-04-01'

# ── 1. get_gen_fuel_type ─────────────────────────────────────────────────────
# GenMixHourly → Hydro, Nuclear, NaturalGas only; forecast = 1-day shift of actual
df1 = ft.get_gen_fuel_type(D, D, var_spec=['a', 'f']) 
print('1. get_gen_fuel_type:', df1.shape)
print('   columns:', df1.columns.tolist())
display(df1)

# ── 2. get_gen_fuel_actual ───────────────────────────────────────────────────
# GenMixHourly → all 10 fuel types; actual only, melted format
df2 = ft.get_gen_fuel_actual(D, D)
print('2. get_gen_fuel_actual:', df2.shape)
print('   fuel types:', df2['fuel_type'].unique().tolist() if 'fuel_type' in df2.columns else df2.columns.tolist())
display(df2)

# ── 3. get_baa_zonal_gen_fuel_type ───────────────────────────────────────────
# BAAZonalGenMixHourly → Hydro, Nuclear, NaturalGas; E and W; forecast = 1-day shift per zone
df3 = ft.get_baa_zonal_gen_fuel_type(D, D, baa_zones=['E', 'W'], var_spec=['a', 'f'])
print('3. get_baa_zonal_gen_fuel_type:', df3.shape)
print('   columns:', df3.columns.tolist())
display(df3)

# ── 4. get_baa_zonal_gen_fuel_actual ─────────────────────────────────────────
# BAAZonalGenMixHourly → all 10 fuel types; E and W; actual only, melted format
df4 = ft.get_baa_zonal_gen_fuel_actual(D, D, baa_zones=['E', 'W'])
print('4. get_baa_zonal_gen_fuel_actual:', df4.shape)
print('   fuel types:', df4['fuel_type'].unique().tolist() if 'fuel_type' in df4.columns else df4.columns.tolist())
display(df4)

# ── 5. get_spp_gas_output (deprecated) ───────────────────────────────────────
# GenMixHourly → NaturalGasTotal only; actual + forecast (1-day shift)
df5 = get_spp_gas_output(D, D, var_spec=['a', 'f'])
print('5. get_spp_gas_output (deprecated):', df5.shape)
print('   columns:', df5.columns.tolist())
display(df5)

1. get_gen_fuel_type: (24, 8)
   columns: ['dt', 'hr', 'spp_fuelmix_hydro_forecast_f', 'spp_fuelmix_nuclear_forecast_f', 'spp_fuelmix_naturalgas_forecast_f', 'spp_fuelmix_hydro_actual_a', 'spp_fuelmix_nuclear_actual_a', 'spp_fuelmix_naturalgas_actual_a']


,dt,hr,spp_fuelmix_hydro_forecast_f,spp_fuelmix_nuclear_forecast_f,spp_fuelmix_naturalgas_forecast_f,spp_fuelmix_hydro_actual_a,spp_fuelmix_nuclear_actual_a,spp_fuelmix_naturalgas_actual_a
0,2026-04-01,1,NaN,NaN,NaN,NaN,NaN,NaN
1,2026-04-01,2,NaN,NaN,NaN,NaN,NaN,NaN
2,2026-04-01,3,NaN,NaN,NaN,NaN,NaN,NaN
3,2026-04-01,4,NaN,NaN,NaN,NaN,NaN,NaN
4,2026-04-01,5,NaN,NaN,NaN,NaN,NaN,NaN
5,2026-04-01,6,NaN,NaN,NaN,NaN,NaN,NaN
6,2026-04-01,7,NaN,NaN,NaN,NaN,NaN,NaN
7,2026-04-01,8,NaN,NaN,NaN,NaN,NaN,NaN
8,2026-04-01,9,NaN,NaN,NaN,NaN,NaN,NaN
9,2026-04-01,10,NaN,NaN,NaN,1637.7,2022.2,14074.2


2. get_gen_fuel_actual: (15, 12)
   fuel types: ['dt', 'hr', 'spp_gen_coal_mw', 'spp_gen_dieselfueloil_mw', 'spp_gen_hydro_mw', 'spp_gen_naturalgas_mw', 'spp_gen_nuclear_mw', 'spp_gen_other_mw', 'spp_gen_solar_mw', 'spp_gen_wastedisposalservices_mw', 'spp_gen_wasteheat_mw', 'spp_gen_wind_mw']


,dt,hr,spp_gen_coal_mw,spp_gen_dieselfueloil_mw,spp_gen_hydro_mw,spp_gen_naturalgas_mw,spp_gen_nuclear_mw,spp_gen_other_mw,spp_gen_solar_mw,spp_gen_wastedisposalservices_mw,spp_gen_wasteheat_mw,spp_gen_wind_mw
0,2026-04-01,10,8277.2,70.2,1637.7,14074.2,2022.2,81.7,772.8,7.0,0.0,7332.9
1,2026-04-01,11,8448.3,124.1,1684.1,14188.4,2023.1,117.3,963.5,7.8,0.0,7052.8
2,2026-04-01,12,8019.9,190.0,1602.4,13116.7,2020.8,42.8,1152.7,7.8,0.0,7547.7
3,2026-04-01,13,7712.1,229.9,1659.3,11788.0,2022.6,44.7,1247.4,7.7,0.0,9298.3
4,2026-04-01,14,7478.9,218.9,1743.4,11707.5,2022.8,57.2,1212.8,7.8,0.0,11275.6
5,2026-04-01,15,7150.5,165.4,1751.6,9968.8,2022.0,70.9,991.6,7.8,0.0,13926.5
6,2026-04-01,16,7259.8,122.4,1780.2,9644.6,2021.9,68.2,854.0,7.9,0.0,15508.1
7,2026-04-01,17,7674.2,72.5,1968.5,9400.1,2021.4,56.2,607.3,7.9,0.0,15249.4
8,2026-04-01,18,7579.6,27.8,2009.2,9109.9,2022.3,50.1,361.8,7.9,0.0,15471.2
9,2026-04-01,19,7865.9,40.8,2102.5,9192.6,2021.2,56.8,118.0,6.7,0.0,15949.2


3. get_baa_zonal_gen_fuel_type: (48, 9)
   columns: ['dt', 'hr', 'baa_zone', 'spp_baa_zonal_fuelmix_hydro_actual_a', 'spp_baa_zonal_fuelmix_hydro_forecast_f', 'spp_baa_zonal_fuelmix_nuclear_actual_a', 'spp_baa_zonal_fuelmix_nuclear_forecast_f', 'spp_baa_zonal_fuelmix_naturalgas_actual_a', 'spp_baa_zonal_fuelmix_naturalgas_forecast_f']


,dt,hr,baa_zone,spp_baa_zonal_fuelmix_hydro_actual_a,spp_baa_zonal_fuelmix_hydro_forecast_f,spp_baa_zonal_fuelmix_nuclear_actual_a,spp_baa_zonal_fuelmix_nuclear_forecast_f,spp_baa_zonal_fuelmix_naturalgas_actual_a,spp_baa_zonal_fuelmix_naturalgas_forecast_f
0,2026-04-01,1,E,NaN,NaN,NaN,NaN,NaN,NaN
1,2026-04-01,1,W,NaN,NaN,NaN,NaN,NaN,NaN
2,2026-04-01,2,E,NaN,NaN,NaN,NaN,NaN,NaN
3,2026-04-01,2,W,NaN,NaN,NaN,NaN,NaN,NaN
4,2026-04-01,3,E,NaN,NaN,NaN,NaN,NaN,NaN
5,2026-04-01,3,W,NaN,NaN,NaN,NaN,NaN,NaN
6,2026-04-01,4,E,NaN,NaN,NaN,NaN,NaN,NaN
7,2026-04-01,4,W,NaN,NaN,NaN,NaN,NaN,NaN
8,2026-04-01,5,E,NaN,NaN,NaN,NaN,NaN,NaN
9,2026-04-01,5,W,NaN,NaN,NaN,NaN,NaN,NaN


4. get_baa_zonal_gen_fuel_actual: (15, 22)
   fuel types: ['dt', 'hr', 'e_spp_baa_zonal_gen_coal_mw', 'e_spp_baa_zonal_gen_dieselfueloil_mw', 'e_spp_baa_zonal_gen_hydro_mw', 'e_spp_baa_zonal_gen_naturalgas_mw', 'e_spp_baa_zonal_gen_nuclear_mw', 'e_spp_baa_zonal_gen_other_mw', 'e_spp_baa_zonal_gen_solar_mw', 'e_spp_baa_zonal_gen_wastedisposalservices_mw', 'e_spp_baa_zonal_gen_wasteheat_mw', 'e_spp_baa_zonal_gen_wind_mw', 'w_spp_baa_zonal_gen_coal_mw', 'w_spp_baa_zonal_gen_dieselfueloil_mw', 'w_spp_baa_zonal_gen_hydro_mw', 'w_spp_baa_zonal_gen_naturalgas_mw', 'w_spp_baa_zonal_gen_nuclear_mw', 'w_spp_baa_zonal_gen_other_mw', 'w_spp_baa_zonal_gen_solar_mw', 'w_spp_baa_zonal_gen_wastedisposalservices_mw', 'w_spp_baa_zonal_gen_wasteheat_mw', 'w_spp_baa_zonal_gen_wind_mw']


,dt,hr,e_spp_baa_zonal_gen_coal_mw,e_spp_baa_zonal_gen_dieselfueloil_mw,e_spp_baa_zonal_gen_hydro_mw,e_spp_baa_zonal_gen_naturalgas_mw,e_spp_baa_zonal_gen_nuclear_mw,e_spp_baa_zonal_gen_other_mw,e_spp_baa_zonal_gen_solar_mw,e_spp_baa_zonal_gen_wastedisposalservices_mw,...,w_spp_baa_zonal_gen_coal_mw,w_spp_baa_zonal_gen_dieselfueloil_mw,w_spp_baa_zonal_gen_hydro_mw,w_spp_baa_zonal_gen_naturalgas_mw,w_spp_baa_zonal_gen_nuclear_mw,w_spp_baa_zonal_gen_other_mw,w_spp_baa_zonal_gen_solar_mw,w_spp_baa_zonal_gen_wastedisposalservices_mw,w_spp_baa_zonal_gen_wasteheat_mw,w_spp_baa_zonal_gen_wind_mw
0,2026-04-01,10,7352.8,1.6,1145.2,13432.7,2022.2,77.0,633.7,7.0,...,924.4,68.6,492.5,641.5,0.0,4.7,139.2,0.0,0.0,597.4
1,2026-04-01,11,7422.4,45.5,1248.4,13560.0,2023.1,111.8,771.1,7.8,...,1025.9,78.6,435.7,628.3,0.0,5.6,192.4,0.0,0.0,581.4
2,2026-04-01,12,7193.6,144.7,1171.2,12634.2,2020.8,38.2,798.1,7.8,...,826.3,45.2,431.2,482.6,0.0,4.6,354.6,0.0,0.0,585.5
3,2026-04-01,13,6943.8,147.6,1222.9,11442.4,2022.6,39.9,882.9,7.7,...,768.3,82.4,436.4,345.6,0.0,4.8,364.5,0.0,0.0,543.9
4,2026-04-01,14,6652.1,147.5,1287.3,11296.4,2022.8,52.6,825.8,7.8,...,826.8,71.3,456.1,411.0,0.0,4.7,387.0,0.0,0.0,526.6
5,2026-04-01,15,6304.0,141.7,1305.1,9643.5,2022.0,66.2,569.6,7.8,...,846.4,23.7,446.5,325.3,0.0,4.7,422.0,0.0,0.0,532.3
6,2026-04-01,16,6472.7,96.4,1316.9,9317.1,2021.9,63.5,439.5,7.9,...,787.1,26.0,463.3,327.6,0.0,4.7,414.4,0.0,0.0,514.8
7,2026-04-01,17,6749.8,44.8,1488.5,8994.1,2021.4,51.5,256.2,7.9,...,924.5,27.7,480.0,406.0,0.0,4.7,351.1,0.0,0.0,486.7
8,2026-04-01,18,6593.0,1.6,1500.2,8658.2,2022.3,45.4,157.5,7.9,...,986.6,26.2,508.9,451.7,0.0,4.7,204.4,0.0,0.0,433.2
9,2026-04-01,19,6809.4,1.5,1507.8,8650.7,2021.2,52.1,49.4,6.7,...,1056.5,39.3,594.7,541.9,0.0,4.7,68.6,0.0,0.0,410.0


5. get_spp_gas_output (deprecated): (24, 4)
   columns: ['dt', 'hr', 'spp_fuelmix_naturalgas_actual_a', 'spp_fuelmix_naturalgas_forecast_f']


/tmp/ipykernel_2635545/3805136757.py:39: DeprecationWarning: Call to deprecated function get_spp_gas_output.
  df5 = get_spp_gas_output(D, D, var_spec=['a', 'f'])


,dt,hr,spp_fuelmix_naturalgas_actual_a,spp_fuelmix_naturalgas_forecast_f
0,2026-04-01,1,NaN,NaN
1,2026-04-01,2,NaN,NaN
2,2026-04-01,3,NaN,NaN
3,2026-04-01,4,NaN,NaN
4,2026-04-01,5,NaN,NaN
5,2026-04-01,6,NaN,NaN
6,2026-04-01,7,NaN,NaN
7,2026-04-01,8,NaN,NaN
8,2026-04-01,9,NaN,NaN
9,2026-04-01,10,NaN,NaN


### 21. Gen Outage by Fuel Type

In [3]:
df_genout_total = genout.get_genoutage_by_level(START, END, area_list=['SPP'], var_spec=['f', 'a'])
print(df_genout_total.shape)
df_genout_total

(24, 5)


,dt,hr,baa_zone,spp_genoutage_forecast_f,spp_genoutage_actual_a
0,2026-04-01,1,SPP,25897.1,25171.9
1,2026-04-01,2,SPP,25907.1,25179.9
2,2026-04-01,3,SPP,25907.1,25180.5
3,2026-04-01,4,SPP,25907.7,25183.5
4,2026-04-01,5,SPP,25182.9,25794.4
5,2026-04-01,6,SPP,25180.6,25792.1
6,2026-04-01,7,SPP,25258.3,25888.2
7,2026-04-01,8,SPP,25152.3,25286.2
8,2026-04-01,9,SPP,25272.4,25314.3
9,2026-04-01,10,SPP,25265.0,25306.9


In [4]:
df_outage_fuel = fueltype.get_genoutage_fuel_type('2026-04-01', '2026-04-01', var_spec=['f', 'a'])
print(df_outage_fuel.shape)

df_outage_fuel['total_f'] = df_outage_fuel.filter(regex='_f$').sum(axis=1)
df_outage_fuel['total_a'] = df_outage_fuel.filter(regex='_a$').sum(axis=1)
df_outage_fuel

(24, 14)


,dt,hr,spp_fuelmixoutage_coal_forecast_f,spp_fuelmixoutage_dieselfueloil_forecast_f,spp_fuelmixoutage_hydro_forecast_f,spp_fuelmixoutage_naturalgas_forecast_f,spp_fuelmixoutage_solar_forecast_f,spp_fuelmixoutage_wind_forecast_f,spp_fuelmixoutage_coal_actual_a,spp_fuelmixoutage_dieselfueloil_actual_a,spp_fuelmixoutage_hydro_actual_a,spp_fuelmixoutage_naturalgas_actual_a,spp_fuelmixoutage_solar_actual_a,spp_fuelmixoutage_wind_actual_a,total_f,total_a
0,2026-04-01,1,8488.8,72.0,936.0,14325.5,1155.7,625.5,8488.8,72.0,890.0,13752.5,1003.5,705.5,25603.5,24912.3
1,2026-04-01,2,8488.8,72.0,936.0,14335.5,1155.7,625.5,8488.8,72.0,890.0,13762.5,1003.5,705.5,25613.5,24922.3
2,2026-04-01,3,8488.8,72.0,936.0,14335.5,1155.7,625.5,8488.8,72.0,890.0,13762.5,1003.5,705.5,25613.5,24922.3
3,2026-04-01,4,8488.8,72.0,936.0,14335.5,1155.7,625.5,8488.8,72.0,890.0,13762.5,1003.5,705.5,25613.5,24922.3
4,2026-04-01,5,8488.8,72.0,890.0,13762.5,1003.5,705.5,8938.8,72.0,890.0,13973.9,781.5,824.5,24922.3,25480.7
5,2026-04-01,6,8488.8,72.0,890.0,13762.5,1003.5,705.5,8938.8,72.0,890.0,13973.9,781.5,824.5,24922.3,25480.7
6,2026-04-01,7,8488.8,72.0,890.0,13847.5,1003.5,705.5,8938.8,72.0,890.0,14077.2,781.5,824.5,25007.3,25584.0
7,2026-04-01,8,8488.8,72.0,890.0,13892.5,853.5,705.5,8488.8,72.0,890.0,14074.2,631.5,824.5,24902.3,24981.0
8,2026-04-01,9,8488.8,266.0,890.0,13892.5,781.5,705.5,8488.8,266.0,890.0,14109.9,631.5,625.5,25024.3,25011.7
9,2026-04-01,10,8488.8,266.0,890.0,13892.5,781.5,705.5,8488.8,266.0,890.0,14109.9,631.5,625.5,25024.3,25011.7


In [5]:
# ── DIAGNOSTIC: why fuel-type sum != GeneratorOutageNew total ──────────────
import pandas as pd
from nighthawk.util import connections, sql_functions

# 1. Raw GeneratorOutageNew — dt/hr already in SPP market time
df_gon = sql_functions.download_df_from_sql_db(
    "SELECT dt, hr, ForecastOutage, ActualOutage "
    "FROM spp_physical.GeneratorOutageNew "
    "WHERE dt = '2026-04-01' ORDER BY hr"
)
df_gon['dt'] = df_gon['dt'].astype(str)
print('GeneratorOutageNew columns:', df_gon.columns.tolist())
print('Rows:', len(df_gon))
display(df_gon)

# 2. Raw fuel-type tables — dthr is UTC hour-ending
#    nighthawk converts: tz_localize(UTC) -> tz_convert(US/Central) - 1h, then hr = local_hour + 1
fuel_cols = ['CoalMW','DieselFuelOilMW','HydroMW','NaturalGasMW','SolarMW','WindMW']

df_ftype_f = sql_functions.download_df_from_sql_db(
    "SELECT dthr, CoalMW, DieselFuelOilMW, HydroMW, NaturalGasMW, SolarMW, WindMW "
    "FROM spp_physical.GenOutagebyFuelTypeForecast "
    "WHERE DATE(dthr) >= DATE_SUB('2026-04-01', INTERVAL 1 DAY) "
    "  AND DATE(dthr) <= DATE_ADD('2026-04-01', INTERVAL 1 DAY) ORDER BY dthr"
)
df_ftype_a = sql_functions.download_df_from_sql_db(
    "SELECT dthr, CoalMW, DieselFuelOilMW, HydroMW, NaturalGasMW, SolarMW, WindMW "
    "FROM spp_physical.GenOutagebyFuelTypeActual "
    "WHERE DATE(dthr) >= DATE_SUB('2026-04-01', INTERVAL 1 DAY) "
    "  AND DATE(dthr) <= DATE_ADD('2026-04-01', INTERVAL 1 DAY) ORDER BY dthr"
)

for df, tag in [(df_ftype_f, 'Forecast'), (df_ftype_a, 'Actual')]:
    df['dthr'] = pd.to_datetime(df['dthr']).dt.tz_localize('UTC').dt.tz_convert('US/Central') - pd.Timedelta('1h')
    df['dt']   = df['dthr'].dt.strftime('%Y-%m-%d')
    df['hr']   = df['dthr'].dt.hour + 1
    df.drop(columns=['dthr'], inplace=True)
    df['fuel_sum'] = df[fuel_cols].sum(axis=1)
    print(f'GenOutagebyFuelType{tag} rows after tz-convert:', len(df))
    display(df[df['dt'] == '2026-04-01'][['dt','hr','fuel_sum'] + fuel_cols])

# 3. Side-by-side on target date
df_f_day = df_ftype_f[df_ftype_f['dt'] == '2026-04-01'][['dt','hr','fuel_sum']].rename(columns={'fuel_sum':'fuel_sum_f'})
df_a_day = df_ftype_a[df_ftype_a['dt'] == '2026-04-01'][['dt','hr','fuel_sum']].rename(columns={'fuel_sum':'fuel_sum_a'})
df_gon_day = df_gon[df_gon['dt'] == '2026-04-01'][['dt','hr','ForecastOutage','ActualOutage']]

cmp = df_gon_day.merge(df_f_day, on=['dt','hr'], how='outer').merge(df_a_day, on=['dt','hr'], how='outer')
cmp['diff_f'] = cmp['ForecastOutage'] - cmp['fuel_sum_f']
cmp['diff_a'] = cmp['ActualOutage']   - cmp['fuel_sum_a']
print('\n=== Comparison: GeneratorOutageNew vs fuel_type_sum ===')
display(cmp)
print('Max |diff_f|:', cmp['diff_f'].abs().max(), '  Max |diff_a|:', cmp['diff_a'].abs().max())
print('Hours with diff_f != 0:', (cmp['diff_f'].abs() > 0.01).sum())
print('Hours with diff_a != 0:', (cmp['diff_a'].abs() > 0.01).sum())

# 4. Inspect all columns of GeneratorOutageNew (may reveal extra fuel type fields)
df_raw = sql_functions.download_df_from_sql_db("SELECT * FROM spp_physical.GeneratorOutageNew WHERE dt = '2026-04-01' LIMIT 3")
print('\nAll columns in GeneratorOutageNew:', df_raw.columns.tolist())
display(df_raw)


GeneratorOutageNew columns: ['dt', 'hr', 'ForecastOutage', 'ActualOutage']
Rows: 24


,dt,hr,ForecastOutage,ActualOutage
0,2026-04-01,1,25715.9,25197.9
1,2026-04-01,2,25728.3,25198.1
2,2026-04-01,3,25878.9,25348.7
3,2026-04-01,4,25878.9,25348.7
4,2026-04-01,5,25712.9,25182.7
5,2026-04-01,6,25612.9,25182.7
6,2026-04-01,7,25897.1,25171.9
7,2026-04-01,8,25907.1,25179.9
8,2026-04-01,9,25907.1,25180.5
9,2026-04-01,10,25907.7,25183.5


GenOutagebyFuelTypeForecast rows after tz-convert: 72


,dt,hr,fuel_sum,CoalMW,DieselFuelOilMW,HydroMW,NaturalGasMW,SolarMW,WindMW
30,2026-04-01,1,25603.5,8488.8,72.0,936.0,14325.5,1155.7,625.5
31,2026-04-01,2,25613.5,8488.8,72.0,936.0,14335.5,1155.7,625.5
32,2026-04-01,3,25613.5,8488.8,72.0,936.0,14335.5,1155.7,625.5
33,2026-04-01,4,25613.5,8488.8,72.0,936.0,14335.5,1155.7,625.5
34,2026-04-01,5,24922.3,8488.8,72.0,890.0,13762.5,1003.5,705.5
35,2026-04-01,6,24922.3,8488.8,72.0,890.0,13762.5,1003.5,705.5
36,2026-04-01,7,25007.3,8488.8,72.0,890.0,13847.5,1003.5,705.5
37,2026-04-01,8,24902.3,8488.8,72.0,890.0,13892.5,853.5,705.5
38,2026-04-01,9,25024.3,8488.8,266.0,890.0,13892.5,781.5,705.5
39,2026-04-01,10,25024.3,8488.8,266.0,890.0,13892.5,781.5,705.5


GenOutagebyFuelTypeActual rows after tz-convert: 72


,dt,hr,fuel_sum,CoalMW,DieselFuelOilMW,HydroMW,NaturalGasMW,SolarMW,WindMW
30,2026-04-01,1,24912.3,8488.8,72.0,890.0,13752.5,1003.5,705.5
31,2026-04-01,2,24922.3,8488.8,72.0,890.0,13762.5,1003.5,705.5
32,2026-04-01,3,24922.3,8488.8,72.0,890.0,13762.5,1003.5,705.5
33,2026-04-01,4,24922.3,8488.8,72.0,890.0,13762.5,1003.5,705.5
34,2026-04-01,5,25480.7,8938.8,72.0,890.0,13973.9,781.5,824.5
35,2026-04-01,6,25480.7,8938.8,72.0,890.0,13973.9,781.5,824.5
36,2026-04-01,7,25584.0,8938.8,72.0,890.0,14077.2,781.5,824.5
37,2026-04-01,8,24981.0,8488.8,72.0,890.0,14074.2,631.5,824.5
38,2026-04-01,9,25011.7,8488.8,266.0,890.0,14109.9,631.5,625.5
39,2026-04-01,10,25011.7,8488.8,266.0,890.0,14109.9,631.5,625.5



=== Comparison: GeneratorOutageNew vs fuel_type_sum ===


,dt,hr,ForecastOutage,ActualOutage,fuel_sum_f,fuel_sum_a,diff_f,diff_a
0,2026-04-01,1,25715.9,25197.9,25603.5,24912.3,112.4,285.6
1,2026-04-01,2,25728.3,25198.1,25613.5,24922.3,114.8,275.8
2,2026-04-01,3,25878.9,25348.7,25613.5,24922.3,265.4,426.4
3,2026-04-01,4,25878.9,25348.7,25613.5,24922.3,265.4,426.4
4,2026-04-01,5,25712.9,25182.7,24922.3,25480.7,790.6,-298.0
5,2026-04-01,6,25612.9,25182.7,24922.3,25480.7,690.6,-298.0
6,2026-04-01,7,25897.1,25171.9,25007.3,25584.0,889.8,-412.1
7,2026-04-01,8,25907.1,25179.9,24902.3,24981.0,1004.8,198.9
8,2026-04-01,9,25907.1,25180.5,25024.3,25011.7,882.8,168.8
9,2026-04-01,10,25907.7,25183.5,25024.3,25011.7,883.4,171.8


Max |diff_f|: 1004.7999999999993   Max |diff_a|: 782.7000000000044
Hours with diff_f != 0: 24
Hours with diff_a != 0: 24

All columns in GeneratorOutageNew: ['dt', 'hr', 'ActualOutage', 'ForecastOutage']


,dt,hr,ActualOutage,ForecastOutage
0,2026-04-01,1,25197.9,25715.9
1,2026-04-01,2,25198.1,25728.3
2,2026-04-01,3,25348.7,25878.9


### 21b. get_genoutage_fuel_type — market-total outage by fuel type

SPP only. Reads `spp_physical.GenOutagebyFuelTypeForecast` and `GenOutagebyFuelTypeActual`.

In [7]:
df_outage_fuel = fueltype.get_genoutage_fuel_type('2026-04-01', '2026-04-01', var_spec=['f', 'a'])
print(df_outage_fuel)

            dt  hr  spp_fuelmixoutage_coal_forecast_f  \
0   2026-04-01   1                             8488.8   
1   2026-04-01   2                             8488.8   
2   2026-04-01   3                             8488.8   
3   2026-04-01   4                             8488.8   
4   2026-04-01   5                             8488.8   
5   2026-04-01   6                             8488.8   
6   2026-04-01   7                             8488.8   
7   2026-04-01   8                             8488.8   
8   2026-04-01   9                             8488.8   
9   2026-04-01  10                             8488.8   
10  2026-04-01  11                             8488.8   
11  2026-04-01  12                             8488.8   
12  2026-04-01  13                             9024.8   
13  2026-04-01  14                             9024.8   
14  2026-04-01  15                             9024.8   
15  2026-04-01  16                             9024.8   
16  2026-04-01  17             

### 21c. get_baa_zonal_genoutage_fuel_type — BAA-zonal outage by fuel type

SPP only. Reads `spp_physical.BAAZonalGenOutagebyFuelTypeForecast` and `BAAZonalGenOutagebyFuelTypeActual`.

In [8]:
df_baa_outage_fuel = fueltype.get_baa_zonal_genoutage_fuel_type('2026-04-01', '2026-04-01', baa_zones=['E', 'W'], var_spec=['f', 'a'])
print(df_baa_outage_fuel.shape)
print(df_baa_outage_fuel.columns.tolist())
display(df_baa_outage_fuel)

(48, 15)
['dt', 'hr', 'baa_zone', 'spp_baa_zonal_fuelmixoutage_coal_forecast_f', 'spp_baa_zonal_fuelmixoutage_dieselfueloil_forecast_f', 'spp_baa_zonal_fuelmixoutage_hydro_forecast_f', 'spp_baa_zonal_fuelmixoutage_naturalgas_forecast_f', 'spp_baa_zonal_fuelmixoutage_solar_forecast_f', 'spp_baa_zonal_fuelmixoutage_wind_forecast_f', 'spp_baa_zonal_fuelmixoutage_coal_actual_a', 'spp_baa_zonal_fuelmixoutage_dieselfueloil_actual_a', 'spp_baa_zonal_fuelmixoutage_hydro_actual_a', 'spp_baa_zonal_fuelmixoutage_naturalgas_actual_a', 'spp_baa_zonal_fuelmixoutage_solar_actual_a', 'spp_baa_zonal_fuelmixoutage_wind_actual_a']


,dt,hr,baa_zone,spp_baa_zonal_fuelmixoutage_coal_forecast_f,spp_baa_zonal_fuelmixoutage_dieselfueloil_forecast_f,spp_baa_zonal_fuelmixoutage_hydro_forecast_f,spp_baa_zonal_fuelmixoutage_naturalgas_forecast_f,spp_baa_zonal_fuelmixoutage_solar_forecast_f,spp_baa_zonal_fuelmixoutage_wind_forecast_f,spp_baa_zonal_fuelmixoutage_coal_actual_a,spp_baa_zonal_fuelmixoutage_dieselfueloil_actual_a,spp_baa_zonal_fuelmixoutage_hydro_actual_a,spp_baa_zonal_fuelmixoutage_naturalgas_actual_a,spp_baa_zonal_fuelmixoutage_solar_actual_a,spp_baa_zonal_fuelmixoutage_wind_actual_a
0,2026-04-01,1,E,8488.8,72.0,936.0,14325.5,1155.7,625.5,8488.8,72.0,890.0,13752.5,1003.5,705.5
1,2026-04-01,1,W,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2026-04-01,2,E,8488.8,72.0,936.0,14335.5,1155.7,625.5,8488.8,72.0,890.0,13762.5,1003.5,705.5
3,2026-04-01,2,W,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2026-04-01,3,E,8488.8,72.0,936.0,14335.5,1155.7,625.5,8488.8,72.0,890.0,13762.5,1003.5,705.5
5,2026-04-01,3,W,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,2026-04-01,4,E,8488.8,72.0,936.0,14335.5,1155.7,625.5,8488.8,72.0,890.0,13762.5,1003.5,705.5
7,2026-04-01,4,W,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8,2026-04-01,5,E,8488.8,72.0,890.0,13762.5,1003.5,705.5,8938.8,72.0,890.0,13973.9,781.5,824.5
9,2026-04-01,5,W,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### 21d. get_data_and_mapping_for_genoutage_fuel_type — wrapper with node mapping

Same as `get_genoutage_fuel_type` but returns `(data, mapping)` keyed by node.

In [9]:
from nighthawk.data.pipeline.var_handler.fuel_type import get_data_and_mapping_for_genoutage_fuel_type

data_goft, mapping_goft = get_data_and_mapping_for_genoutage_fuel_type(
    node_list=[636], opexchange='SPP',
    start_dt='2026-04-01', end_dt='2026-04-01',
    fuel_type=['Coal', 'DieselFuelOil', 'Hydro', 'NaturalGas', 'Solar', 'Wind'],
    impute=True
)
print('data shape:', data_goft.shape)
print('mapping shape:', mapping_goft.shape)
display(data_goft)
display(mapping_goft)

             dt  hr  spp_fuelmixoutage_coal_forecast_f  \
0 0  2026-04-01   1                             8488.8   
1 1  2026-04-01   2                             8488.8   
2 2  2026-04-01   3                             8488.8   
3 3  2026-04-01   4                             8488.8   
4 4  2026-04-01   5                             8488.8   

     spp_fuelmixoutage_dieselfueloil_forecast_f  \
0 0                                        72.0   
1 1                                        72.0   
2 2                                        72.0   
3 3                                        72.0   
4 4                                        72.0   

     spp_fuelmixoutage_hydro_forecast_f  \
0 0                               936.0   
1 1                               936.0   
2 2                               936.0   
3 3                               936.0   
4 4                               890.0   

     spp_fuelmixoutage_naturalgas_forecast_f  \
0 0                                  

,,dt,hr,spp_fuelmixoutage_coal_forecast_f,spp_fuelmixoutage_coal_actual_a,spp_fuelmixoutage_dieselfueloil_forecast_f,spp_fuelmixoutage_dieselfueloil_actual_a,spp_fuelmixoutage_hydro_forecast_f,spp_fuelmixoutage_hydro_actual_a,spp_fuelmixoutage_naturalgas_forecast_f,spp_fuelmixoutage_naturalgas_actual_a,spp_fuelmixoutage_solar_forecast_f,spp_fuelmixoutage_solar_actual_a,spp_fuelmixoutage_wind_forecast_f,spp_fuelmixoutage_wind_actual_a
0,0,2026-04-01,1,8488.8,8488.8,72.0,72.0,936.0,890.0,14325.5,13752.5,1155.7,1003.5,625.5,705.5
1,1,2026-04-01,2,8488.8,8488.8,72.0,72.0,936.0,890.0,14335.5,13762.5,1155.7,1003.5,625.5,705.5
2,2,2026-04-01,3,8488.8,8488.8,72.0,72.0,936.0,890.0,14335.5,13762.5,1155.7,1003.5,625.5,705.5
3,3,2026-04-01,4,8488.8,8488.8,72.0,72.0,936.0,890.0,14335.5,13762.5,1155.7,1003.5,625.5,705.5
4,4,2026-04-01,5,8488.8,8938.8,72.0,72.0,890.0,890.0,13762.5,13973.9,1003.5,781.5,705.5,824.5
5,5,2026-04-01,6,8488.8,8938.8,72.0,72.0,890.0,890.0,13762.5,13973.9,1003.5,781.5,705.5,824.5
6,6,2026-04-01,7,8488.8,8938.8,72.0,72.0,890.0,890.0,13847.5,14077.2,1003.5,781.5,705.5,824.5
7,7,2026-04-01,8,8488.8,8488.8,72.0,72.0,890.0,890.0,13892.5,14074.2,853.5,631.5,705.5,824.5
8,8,2026-04-01,9,8488.8,8488.8,266.0,266.0,890.0,890.0,13892.5,14109.9,781.5,631.5,705.5,625.5
9,9,2026-04-01,10,8488.8,8488.8,266.0,266.0,890.0,890.0,13892.5,14109.9,781.5,631.5,705.5,625.5


,node_num,var_name
0,636,spp_fuelmixoutage_coal_forecast_f
1,636,spp_fuelmixoutage_coal_actual_a
2,636,spp_fuelmixoutage_dieselfueloil_forecast_f
3,636,spp_fuelmixoutage_dieselfueloil_actual_a
4,636,spp_fuelmixoutage_hydro_forecast_f
5,636,spp_fuelmixoutage_hydro_actual_a
6,636,spp_fuelmixoutage_naturalgas_forecast_f
7,636,spp_fuelmixoutage_naturalgas_actual_a
8,636,spp_fuelmixoutage_solar_forecast_f
9,636,spp_fuelmixoutage_solar_actual_a


### 21e. get_data_and_mapping_for_baa_zonal_genoutage_fuel_type — BAA-zonal wrapper with node mapping

In [10]:
from nighthawk.data.pipeline.var_handler.fuel_type import get_data_and_mapping_for_baa_zonal_genoutage_fuel_type

data_bgoft, mapping_bgoft = get_data_and_mapping_for_baa_zonal_genoutage_fuel_type(
    node_list=[636], opexchange='SPP',
    start_dt='2026-04-01', end_dt='2026-04-01',
    fuel_type=['All'],
    impute=True
)
print('data shape:', data_bgoft.shape)
print('mapping shape:', mapping_bgoft.shape)
display(data_bgoft)
display(mapping_bgoft)

data shape: (24, 14)
mapping shape: (12, 2)


,dt,hr,e_spp_baa_zonal_fuelmixoutage_coal_actual_a,e_spp_baa_zonal_fuelmixoutage_coal_forecast_f,e_spp_baa_zonal_fuelmixoutage_dieselfueloil_actual_a,e_spp_baa_zonal_fuelmixoutage_dieselfueloil_forecast_f,e_spp_baa_zonal_fuelmixoutage_hydro_actual_a,e_spp_baa_zonal_fuelmixoutage_hydro_forecast_f,e_spp_baa_zonal_fuelmixoutage_naturalgas_actual_a,e_spp_baa_zonal_fuelmixoutage_naturalgas_forecast_f,e_spp_baa_zonal_fuelmixoutage_solar_actual_a,e_spp_baa_zonal_fuelmixoutage_solar_forecast_f,e_spp_baa_zonal_fuelmixoutage_wind_actual_a,e_spp_baa_zonal_fuelmixoutage_wind_forecast_f
0,2026-04-01,1,8488.8,8488.8,72.0,72.0,890.0,936.0,13752.5,14325.5,1003.5,1155.7,705.5,625.5
1,2026-04-01,2,8488.8,8488.8,72.0,72.0,890.0,936.0,13762.5,14335.5,1003.5,1155.7,705.5,625.5
2,2026-04-01,3,8488.8,8488.8,72.0,72.0,890.0,936.0,13762.5,14335.5,1003.5,1155.7,705.5,625.5
3,2026-04-01,4,8488.8,8488.8,72.0,72.0,890.0,936.0,13762.5,14335.5,1003.5,1155.7,705.5,625.5
4,2026-04-01,5,8938.8,8488.8,72.0,72.0,890.0,890.0,13973.9,13762.5,781.5,1003.5,824.5,705.5
5,2026-04-01,6,8938.8,8488.8,72.0,72.0,890.0,890.0,13973.9,13762.5,781.5,1003.5,824.5,705.5
6,2026-04-01,7,8938.8,8488.8,72.0,72.0,890.0,890.0,14077.2,13847.5,781.5,1003.5,824.5,705.5
7,2026-04-01,8,8488.8,8488.8,72.0,72.0,890.0,890.0,14074.2,13892.5,631.5,853.5,824.5,705.5
8,2026-04-01,9,8488.8,8488.8,266.0,266.0,890.0,890.0,14109.9,13892.5,631.5,781.5,625.5,705.5
9,2026-04-01,10,8488.8,8488.8,266.0,266.0,890.0,890.0,14109.9,13892.5,631.5,781.5,625.5,705.5


,node_num,var_name
0,636,e_spp_baa_zonal_fuelmixoutage_coal_actual_a
1,636,e_spp_baa_zonal_fuelmixoutage_coal_forecast_f
2,636,e_spp_baa_zonal_fuelmixoutage_dieselfueloil_ac...
3,636,e_spp_baa_zonal_fuelmixoutage_dieselfueloil_fo...
4,636,e_spp_baa_zonal_fuelmixoutage_hydro_actual_a
5,636,e_spp_baa_zonal_fuelmixoutage_hydro_forecast_f
6,636,e_spp_baa_zonal_fuelmixoutage_naturalgas_actual_a
7,636,e_spp_baa_zonal_fuelmixoutage_naturalgas_forec...
8,636,e_spp_baa_zonal_fuelmixoutage_solar_actual_a
9,636,e_spp_baa_zonal_fuelmixoutage_solar_forecast_f


---
## GEN MIX RAW TABLES

### 22. GenMix 5-min (spp_physical.GenMix)

In [9]:
df_genmix = sql_functions.download_df_from_sql_db(
    "SELECT IntervalEnd, GMTIntervalEnd, CoalTotal, WindTotal, SolarTotal, NaturalGasTotal, AverageActualLoad "
    "FROM spp_physical.GenMix "
    "WHERE DATE(IntervalEnd) = '2026-04-03' ORDER BY IntervalEnd")
print(df_genmix.shape)
# spot check: rows should be ~288 (5-min intervals * 24hr)
df_genmix.tail(10)

(108, 7)


,IntervalEnd,GMTIntervalEnd,CoalTotal,WindTotal,SolarTotal,NaturalGasTotal,AverageActualLoad
98,2026-04-03 08:10:00,2026-04-03 13:10:00,5456.2,20804.9,108.1,5611.3,33914.0
99,2026-04-03 08:15:00,2026-04-03 13:15:00,5513.1,20716.8,133.0,5740.2,34431.3
100,2026-04-03 08:20:00,2026-04-03 13:20:00,5541.2,20800.6,174.1,5700.4,34162.5
101,2026-04-03 08:25:00,2026-04-03 13:25:00,5633.1,20924.8,226.1,5633.5,34231.8
102,2026-04-03 08:30:00,2026-04-03 13:30:00,5689.0,20713.1,270.4,5768.1,34704.9
103,2026-04-03 08:35:00,2026-04-03 13:35:00,5768.0,20494.4,262.8,5922.1,34320.7
104,2026-04-03 08:40:00,2026-04-03 13:40:00,5804.2,20453.4,309.1,5830.4,34332.5
105,2026-04-03 08:45:00,2026-04-03 13:45:00,5828.2,20454.6,315.8,5910.4,34833.5
106,2026-04-03 08:50:00,2026-04-03 13:50:00,5830.0,20347.7,471.5,5949.8,34488.7
107,2026-04-03 08:55:00,2026-04-03 13:55:00,5804.3,20208.6,487.8,5969.1,34634.3


### 23. GenMixHourly (spp_physical.GenMixHourly)

In [83]:
df_genmixhr = sql_functions.download_df_from_sql_db(
    "SELECT * "
    "FROM spp_physical.GenMixHourly "
    "WHERE dt = '2026-04-01' ORDER BY hr")
print(df_genmixhr.shape)
pd.set_option('display.max_columns', None)  # show all columns
pd.set_option('display.max_rows', None)     # show all rows
pd.set_option('display.width', None)
display(df_genmixhr)

(14, 33)


,dt,hr,CoalMarket,CoalSelf,CoalTotal,DieselFuelOilMarket,DieselFuelOilSelf,DieselFuelOilTotal,HydroMarket,HydroSelf,HydroTotal,NaturalGasMarket,NaturalGasSelf,NaturalGasTotal,NuclearMarket,NuclearSelf,NuclearTotal,SolarMarket,SolarSelf,SolarTotal,WasteDisposalServicesMarket,WasteDisposalServicesSelf,WasteDisposalServicesTotal,WindMarket,WindSelf,WindTotal,WasteHeatMarket,WasteHeatSelf,WasteHeatTotal,OtherMarket,OtherSelf,OtherTotal,AverageActualLoad
0,2026-04-01,10,3564.0,4713.2,8277.2,0.0,70.2,70.2,79.8,1557.9,1637.7,12165.6,1908.7,14074.2,0.0,2022.2,2022.2,0.0,772.8,772.8,1.5,5.5,7.0,0.0,7332.9,7332.9,0.0,0.0,0.0,43.4,38.2,81.7,35172.4
1,2026-04-01,11,3566.4,4881.9,8448.3,37.6,86.6,124.1,2.4,1681.7,1684.1,12274.8,1913.6,14188.4,0.0,2023.1,2023.1,0.0,963.5,963.5,1.5,6.3,7.8,0.0,7052.8,7052.8,0.0,0.0,0.0,79.2,38.1,117.3,35616.7
2,2026-04-01,12,3497.6,4522.4,8019.9,143.2,46.8,190.0,2.4,1600.0,1602.4,11582.3,1534.5,13116.7,0.0,2020.8,2020.8,0.0,1152.7,1152.7,1.5,6.2,7.8,0.0,7547.7,7547.7,0.0,0.0,0.0,7.3,35.5,42.8,35874.3
3,2026-04-01,13,3324.2,4387.9,7712.1,146.0,83.9,229.9,6.1,1653.2,1659.3,10518.3,1269.7,11788.0,0.0,2022.6,2022.6,0.0,1247.4,1247.4,1.5,6.2,7.7,0.0,9298.3,9298.3,0.0,0.0,0.0,11.6,33.1,44.7,36047.3
4,2026-04-01,14,3221.3,4257.6,7478.9,146.0,72.9,218.9,6.4,1737.0,1743.4,10148.7,1558.7,11707.5,0.0,2022.8,2022.8,0.0,1212.8,1212.8,1.5,6.2,7.8,0.0,11275.6,11275.6,0.0,0.0,0.0,25.4,31.8,57.2,36238.6
5,2026-04-01,15,3119.9,4030.6,7150.5,140.2,25.2,165.4,9.6,1742.0,1751.6,8090.6,1878.1,9968.8,0.0,2022.0,2022.0,0.0,991.6,991.6,1.5,6.2,7.8,0.0,13926.5,13926.5,0.0,0.0,0.0,12.6,58.3,70.9,36344.5
6,2026-04-01,16,3130.6,4129.2,7259.8,94.8,27.6,122.4,20.3,1759.8,1780.2,7719.1,1925.6,9644.6,0.0,2021.9,2021.9,0.0,854.0,854.0,1.5,6.3,7.9,0.0,15508.1,15508.1,0.0,0.0,0.0,38.9,29.2,68.2,36643.9
7,2026-04-01,17,3309.6,4364.6,7674.2,43.3,29.2,72.5,32.6,1935.9,1968.5,8024.9,1375.3,9400.1,0.0,2021.4,2021.4,0.0,607.3,607.3,1.6,6.4,7.9,0.0,15249.4,15249.4,0.0,0.0,0.0,27.0,29.2,56.2,36818.3
8,2026-04-01,18,3238.3,4341.3,7579.6,0.0,27.8,27.8,52.2,1957.0,2009.2,7727.6,1382.3,9109.9,0.0,2022.3,2022.3,0.0,361.8,361.8,1.5,6.4,7.9,0.0,15471.2,15471.2,0.0,0.0,0.0,20.9,29.2,50.1,36795.3
9,2026-04-01,19,3296.8,4569.1,7865.9,0.0,40.8,40.8,52.6,2049.8,2102.5,7704.4,1488.1,9192.6,0.0,2021.2,2021.2,0.0,118.0,118.0,0.4,6.4,6.7,0.0,15949.2,15949.2,0.0,0.0,0.0,27.6,29.2,56.8,36578.6


In [ ]:
df_genmixhr = sql_functions.download_df_from_sql_db(
    "SELECT * "
    "FROM spp_physical.GenMixHourly "
    "WHERE dt = '2026-04-01' ORDER BY hr")
print(df_genmixhr.shape)
pd.set_option('display.max_columns', None)  # show all columns
pd.set_option('display.max_rows', None)     # show all rows
pd.set_option('display.width', None)
display(df_genmixhr)

(14, 33)


,dt,hr,CoalMarket,CoalSelf,CoalTotal,DieselFuelOilMarket,DieselFuelOilSelf,DieselFuelOilTotal,HydroMarket,HydroSelf,HydroTotal,NaturalGasMarket,NaturalGasSelf,NaturalGasTotal,NuclearMarket,NuclearSelf,NuclearTotal,SolarMarket,SolarSelf,SolarTotal,WasteDisposalServicesMarket,WasteDisposalServicesSelf,WasteDisposalServicesTotal,WindMarket,WindSelf,WindTotal,WasteHeatMarket,WasteHeatSelf,WasteHeatTotal,OtherMarket,OtherSelf,OtherTotal,AverageActualLoad
0,2026-04-01,10,3564.0,4713.2,8277.2,0.0,70.2,70.2,79.8,1557.9,1637.7,12165.6,1908.7,14074.2,0.0,2022.2,2022.2,0.0,772.8,772.8,1.5,5.5,7.0,0.0,7332.9,7332.9,0.0,0.0,0.0,43.4,38.2,81.7,35172.4
1,2026-04-01,11,3566.4,4881.9,8448.3,37.6,86.6,124.1,2.4,1681.7,1684.1,12274.8,1913.6,14188.4,0.0,2023.1,2023.1,0.0,963.5,963.5,1.5,6.3,7.8,0.0,7052.8,7052.8,0.0,0.0,0.0,79.2,38.1,117.3,35616.7
2,2026-04-01,12,3497.6,4522.4,8019.9,143.2,46.8,190.0,2.4,1600.0,1602.4,11582.3,1534.5,13116.7,0.0,2020.8,2020.8,0.0,1152.7,1152.7,1.5,6.2,7.8,0.0,7547.7,7547.7,0.0,0.0,0.0,7.3,35.5,42.8,35874.3
3,2026-04-01,13,3324.2,4387.9,7712.1,146.0,83.9,229.9,6.1,1653.2,1659.3,10518.3,1269.7,11788.0,0.0,2022.6,2022.6,0.0,1247.4,1247.4,1.5,6.2,7.7,0.0,9298.3,9298.3,0.0,0.0,0.0,11.6,33.1,44.7,36047.3
4,2026-04-01,14,3221.3,4257.6,7478.9,146.0,72.9,218.9,6.4,1737.0,1743.4,10148.7,1558.7,11707.5,0.0,2022.8,2022.8,0.0,1212.8,1212.8,1.5,6.2,7.8,0.0,11275.6,11275.6,0.0,0.0,0.0,25.4,31.8,57.2,36238.6
5,2026-04-01,15,3119.9,4030.6,7150.5,140.2,25.2,165.4,9.6,1742.0,1751.6,8090.6,1878.1,9968.8,0.0,2022.0,2022.0,0.0,991.6,991.6,1.5,6.2,7.8,0.0,13926.5,13926.5,0.0,0.0,0.0,12.6,58.3,70.9,36344.5
6,2026-04-01,16,3130.6,4129.2,7259.8,94.8,27.6,122.4,20.3,1759.8,1780.2,7719.1,1925.6,9644.6,0.0,2021.9,2021.9,0.0,854.0,854.0,1.5,6.3,7.9,0.0,15508.1,15508.1,0.0,0.0,0.0,38.9,29.2,68.2,36643.9
7,2026-04-01,17,3309.6,4364.6,7674.2,43.3,29.2,72.5,32.6,1935.9,1968.5,8024.9,1375.3,9400.1,0.0,2021.4,2021.4,0.0,607.3,607.3,1.6,6.4,7.9,0.0,15249.4,15249.4,0.0,0.0,0.0,27.0,29.2,56.2,36818.3
8,2026-04-01,18,3238.3,4341.3,7579.6,0.0,27.8,27.8,52.2,1957.0,2009.2,7727.6,1382.3,9109.9,0.0,2022.3,2022.3,0.0,361.8,361.8,1.5,6.4,7.9,0.0,15471.2,15471.2,0.0,0.0,0.0,20.9,29.2,50.1,36795.3
9,2026-04-01,19,3296.8,4569.1,7865.9,0.0,40.8,40.8,52.6,2049.8,2102.5,7704.4,1488.1,9192.6,0.0,2021.2,2021.2,0.0,118.0,118.0,0.4,6.4,6.7,0.0,15949.2,15949.2,0.0,0.0,0.0,27.6,29.2,56.8,36578.6


In [ ]:
df_genmixhr = sql_functions.download_df_from_sql_db(
    "SELECT * "
    "FROM spp_physical.GenMixHourly "
    "WHERE dt = '2026-04-01' ORDER BY hr")
print(df_genmixhr.shape)
pd.set_option('display.max_columns', None)  # show all columns
pd.set_option('display.max_rows', None)     # show all rows
pd.set_option('display.width', None)
display(df_genmixhr)

(14, 33)


,dt,hr,CoalMarket,CoalSelf,CoalTotal,DieselFuelOilMarket,DieselFuelOilSelf,DieselFuelOilTotal,HydroMarket,HydroSelf,HydroTotal,NaturalGasMarket,NaturalGasSelf,NaturalGasTotal,NuclearMarket,NuclearSelf,NuclearTotal,SolarMarket,SolarSelf,SolarTotal,WasteDisposalServicesMarket,WasteDisposalServicesSelf,WasteDisposalServicesTotal,WindMarket,WindSelf,WindTotal,WasteHeatMarket,WasteHeatSelf,WasteHeatTotal,OtherMarket,OtherSelf,OtherTotal,AverageActualLoad
0,2026-04-01,10,3564.0,4713.2,8277.2,0.0,70.2,70.2,79.8,1557.9,1637.7,12165.6,1908.7,14074.2,0.0,2022.2,2022.2,0.0,772.8,772.8,1.5,5.5,7.0,0.0,7332.9,7332.9,0.0,0.0,0.0,43.4,38.2,81.7,35172.4
1,2026-04-01,11,3566.4,4881.9,8448.3,37.6,86.6,124.1,2.4,1681.7,1684.1,12274.8,1913.6,14188.4,0.0,2023.1,2023.1,0.0,963.5,963.5,1.5,6.3,7.8,0.0,7052.8,7052.8,0.0,0.0,0.0,79.2,38.1,117.3,35616.7
2,2026-04-01,12,3497.6,4522.4,8019.9,143.2,46.8,190.0,2.4,1600.0,1602.4,11582.3,1534.5,13116.7,0.0,2020.8,2020.8,0.0,1152.7,1152.7,1.5,6.2,7.8,0.0,7547.7,7547.7,0.0,0.0,0.0,7.3,35.5,42.8,35874.3
3,2026-04-01,13,3324.2,4387.9,7712.1,146.0,83.9,229.9,6.1,1653.2,1659.3,10518.3,1269.7,11788.0,0.0,2022.6,2022.6,0.0,1247.4,1247.4,1.5,6.2,7.7,0.0,9298.3,9298.3,0.0,0.0,0.0,11.6,33.1,44.7,36047.3
4,2026-04-01,14,3221.3,4257.6,7478.9,146.0,72.9,218.9,6.4,1737.0,1743.4,10148.7,1558.7,11707.5,0.0,2022.8,2022.8,0.0,1212.8,1212.8,1.5,6.2,7.8,0.0,11275.6,11275.6,0.0,0.0,0.0,25.4,31.8,57.2,36238.6
5,2026-04-01,15,3119.9,4030.6,7150.5,140.2,25.2,165.4,9.6,1742.0,1751.6,8090.6,1878.1,9968.8,0.0,2022.0,2022.0,0.0,991.6,991.6,1.5,6.2,7.8,0.0,13926.5,13926.5,0.0,0.0,0.0,12.6,58.3,70.9,36344.5
6,2026-04-01,16,3130.6,4129.2,7259.8,94.8,27.6,122.4,20.3,1759.8,1780.2,7719.1,1925.6,9644.6,0.0,2021.9,2021.9,0.0,854.0,854.0,1.5,6.3,7.9,0.0,15508.1,15508.1,0.0,0.0,0.0,38.9,29.2,68.2,36643.9
7,2026-04-01,17,3309.6,4364.6,7674.2,43.3,29.2,72.5,32.6,1935.9,1968.5,8024.9,1375.3,9400.1,0.0,2021.4,2021.4,0.0,607.3,607.3,1.6,6.4,7.9,0.0,15249.4,15249.4,0.0,0.0,0.0,27.0,29.2,56.2,36818.3
8,2026-04-01,18,3238.3,4341.3,7579.6,0.0,27.8,27.8,52.2,1957.0,2009.2,7727.6,1382.3,9109.9,0.0,2022.3,2022.3,0.0,361.8,361.8,1.5,6.4,7.9,0.0,15471.2,15471.2,0.0,0.0,0.0,20.9,29.2,50.1,36795.3
9,2026-04-01,19,3296.8,4569.1,7865.9,0.0,40.8,40.8,52.6,2049.8,2102.5,7704.4,1488.1,9192.6,0.0,2021.2,2021.2,0.0,118.0,118.0,0.4,6.4,6.7,0.0,15949.2,15949.2,0.0,0.0,0.0,27.6,29.2,56.8,36578.6


In [ ]:
df_genmixhr = sql_functions.download_df_from_sql_db(
    "SELECT * "
    "FROM spp_physical.GenMixHourly "
    "WHERE dt = '2026-04-01' ORDER BY hr")
print(df_genmixhr.shape)
pd.set_option('display.max_columns', None)  # show all columns
pd.set_option('display.max_rows', None)     # show all rows
pd.set_option('display.width', None)
display(df_genmixhr)

(14, 33)


,dt,hr,CoalMarket,CoalSelf,CoalTotal,DieselFuelOilMarket,DieselFuelOilSelf,DieselFuelOilTotal,HydroMarket,HydroSelf,HydroTotal,NaturalGasMarket,NaturalGasSelf,NaturalGasTotal,NuclearMarket,NuclearSelf,NuclearTotal,SolarMarket,SolarSelf,SolarTotal,WasteDisposalServicesMarket,WasteDisposalServicesSelf,WasteDisposalServicesTotal,WindMarket,WindSelf,WindTotal,WasteHeatMarket,WasteHeatSelf,WasteHeatTotal,OtherMarket,OtherSelf,OtherTotal,AverageActualLoad
0,2026-04-01,10,3564.0,4713.2,8277.2,0.0,70.2,70.2,79.8,1557.9,1637.7,12165.6,1908.7,14074.2,0.0,2022.2,2022.2,0.0,772.8,772.8,1.5,5.5,7.0,0.0,7332.9,7332.9,0.0,0.0,0.0,43.4,38.2,81.7,35172.4
1,2026-04-01,11,3566.4,4881.9,8448.3,37.6,86.6,124.1,2.4,1681.7,1684.1,12274.8,1913.6,14188.4,0.0,2023.1,2023.1,0.0,963.5,963.5,1.5,6.3,7.8,0.0,7052.8,7052.8,0.0,0.0,0.0,79.2,38.1,117.3,35616.7
2,2026-04-01,12,3497.6,4522.4,8019.9,143.2,46.8,190.0,2.4,1600.0,1602.4,11582.3,1534.5,13116.7,0.0,2020.8,2020.8,0.0,1152.7,1152.7,1.5,6.2,7.8,0.0,7547.7,7547.7,0.0,0.0,0.0,7.3,35.5,42.8,35874.3
3,2026-04-01,13,3324.2,4387.9,7712.1,146.0,83.9,229.9,6.1,1653.2,1659.3,10518.3,1269.7,11788.0,0.0,2022.6,2022.6,0.0,1247.4,1247.4,1.5,6.2,7.7,0.0,9298.3,9298.3,0.0,0.0,0.0,11.6,33.1,44.7,36047.3
4,2026-04-01,14,3221.3,4257.6,7478.9,146.0,72.9,218.9,6.4,1737.0,1743.4,10148.7,1558.7,11707.5,0.0,2022.8,2022.8,0.0,1212.8,1212.8,1.5,6.2,7.8,0.0,11275.6,11275.6,0.0,0.0,0.0,25.4,31.8,57.2,36238.6
5,2026-04-01,15,3119.9,4030.6,7150.5,140.2,25.2,165.4,9.6,1742.0,1751.6,8090.6,1878.1,9968.8,0.0,2022.0,2022.0,0.0,991.6,991.6,1.5,6.2,7.8,0.0,13926.5,13926.5,0.0,0.0,0.0,12.6,58.3,70.9,36344.5
6,2026-04-01,16,3130.6,4129.2,7259.8,94.8,27.6,122.4,20.3,1759.8,1780.2,7719.1,1925.6,9644.6,0.0,2021.9,2021.9,0.0,854.0,854.0,1.5,6.3,7.9,0.0,15508.1,15508.1,0.0,0.0,0.0,38.9,29.2,68.2,36643.9
7,2026-04-01,17,3309.6,4364.6,7674.2,43.3,29.2,72.5,32.6,1935.9,1968.5,8024.9,1375.3,9400.1,0.0,2021.4,2021.4,0.0,607.3,607.3,1.6,6.4,7.9,0.0,15249.4,15249.4,0.0,0.0,0.0,27.0,29.2,56.2,36818.3
8,2026-04-01,18,3238.3,4341.3,7579.6,0.0,27.8,27.8,52.2,1957.0,2009.2,7727.6,1382.3,9109.9,0.0,2022.3,2022.3,0.0,361.8,361.8,1.5,6.4,7.9,0.0,15471.2,15471.2,0.0,0.0,0.0,20.9,29.2,50.1,36795.3
9,2026-04-01,19,3296.8,4569.1,7865.9,0.0,40.8,40.8,52.6,2049.8,2102.5,7704.4,1488.1,9192.6,0.0,2021.2,2021.2,0.0,118.0,118.0,0.4,6.4,6.7,0.0,15949.2,15949.2,0.0,0.0,0.0,27.6,29.2,56.8,36578.6


### 24. BAAZonalGenMix (spp_physical.BAAZonalGenMix) — E and W

In [75]:
df_baa_genmix = sql_functions.download_df_from_sql_db(
    "SELECT IntervalEnd, BAA, CoalTotal, WindTotal, SolarTotal, NaturalGasTotal, AverageActualLoad "
    "FROM spp_physical.BAAZonalGenMix "
    "WHERE DATE(IntervalEnd) = '2026-04-06' ORDER BY IntervalEnd DESC, BAA")
print(df_baa_genmix.shape)
# spot check: E + W rows; WindTotal_E + WindTotal_W should ≈ GenMix WindTotal
df_baa_genmix.head(10)

(264, 7)


,IntervalEnd,BAA,CoalTotal,WindTotal,SolarTotal,NaturalGasTotal,AverageActualLoad
0,2026-04-06 10:55:00,E,4139.9,12061.9,1681.8,8962.5,30808.5
1,2026-04-06 10:55:00,W,619.5,278.6,360.2,802.3,2342.3
2,2026-04-06 10:50:00,E,4200.8,11979.4,1690.8,8980.7,30880.1
3,2026-04-06 10:50:00,W,631.6,273.1,380.9,822.6,2346.4
4,2026-04-06 10:45:00,E,4261.3,11928.1,1659.3,9034.6,31079.2
5,2026-04-06 10:45:00,W,649.3,307.6,446.7,846.5,2771.5
6,2026-04-06 10:40:00,E,4312.2,11886.3,1666.2,9134.4,31005.6
7,2026-04-06 10:40:00,W,636.4,303.5,436.5,846.8,2518.8
8,2026-04-06 10:35:00,E,4389.7,11800.5,1666.4,9189.7,31081.7
9,2026-04-06 10:35:00,W,648.4,318.2,445.9,847.1,2533.6


### 25. BAAZonalGenMixHourly (spp_physical.BAAZonalGenMixHourly) — E + W sum vs GenMixHourly

In [85]:
df_baa_hr = sql_functions.download_df_from_sql_db(
    "SELECT * "
    "FROM spp_physical.BAAZonalGenMixHourly "
    "WHERE dt = '2026-04-01' ORDER BY hr, BAA")
print(df_baa_hr.shape)
# spot check: E + W rows; WindTotal_E + WindTotal_W should ≈ GenMix WindTotal
df_baa_hr.head(10)

(28, 34)


,dt,hr,BAA,CoalMarket,CoalSelf,CoalTotal,DieselFuelOilMarket,DieselFuelOilSelf,DieselFuelOilTotal,HydroMarket,HydroSelf,HydroTotal,NaturalGasMarket,NaturalGasSelf,NaturalGasTotal,NuclearMarket,NuclearSelf,NuclearTotal,SolarMarket,SolarSelf,SolarTotal,WasteDisposalServicesMarket,WasteDisposalServicesSelf,WasteDisposalServicesTotal,WindMarket,WindSelf,WindTotal,WasteHeatMarket,WasteHeatSelf,WasteHeatTotal,OtherMarket,OtherSelf,OtherTotal,AverageActualLoad
0,2026-04-01,10,E,3564.0,3788.8,7352.8,0.0,1.6,1.6,79.8,1065.4,1145.2,12165.6,1267.2,13432.7,0.0,2022.2,2022.2,0.0,633.7,633.7,1.5,5.5,7.0,0.0,6735.6,6735.6,0.0,0.0,0.0,43.4,33.5,77.0,32438.6
1,2026-04-01,10,W,0.0,924.4,924.4,0.0,68.6,68.6,0.0,492.5,492.5,0.0,641.5,641.5,0.0,0.0,0.0,0.0,139.2,139.2,0.0,0.0,0.0,0.0,597.4,597.4,0.0,0.0,0.0,0.0,4.7,4.7,2733.8
2,2026-04-01,11,E,3566.4,3856.0,7422.4,37.6,7.9,45.5,2.4,1246.0,1248.4,12274.8,1285.3,13560.0,0.0,2023.1,2023.1,0.0,771.1,771.1,1.5,6.3,7.8,0.0,6471.4,6471.4,0.0,0.0,0.0,79.2,32.5,111.8,32887.3
3,2026-04-01,11,W,0.0,1025.9,1025.9,0.0,78.6,78.6,0.0,435.7,435.7,0.0,628.3,628.3,0.0,0.0,0.0,0.0,192.4,192.4,0.0,0.0,0.0,0.0,581.4,581.4,0.0,0.0,0.0,0.0,5.6,5.6,2729.4
4,2026-04-01,12,E,3497.6,3696.1,7193.6,143.2,1.6,144.7,2.4,1168.8,1171.2,11582.3,1051.9,12634.2,0.0,2020.8,2020.8,0.0,798.1,798.1,1.5,6.2,7.8,0.0,6962.3,6962.3,0.0,0.0,0.0,7.3,30.9,38.2,33195.4
5,2026-04-01,12,W,0.0,826.3,826.3,0.0,45.2,45.2,0.0,431.2,431.2,0.0,482.6,482.6,0.0,0.0,0.0,0.0,354.6,354.6,0.0,0.0,0.0,0.0,585.5,585.5,0.0,0.0,0.0,0.0,4.6,4.6,2678.9
6,2026-04-01,13,E,3324.2,3619.6,6943.8,146.0,1.6,147.6,6.1,1216.8,1222.9,10518.3,924.2,11442.4,0.0,2022.6,2022.6,0.0,882.9,882.9,1.5,6.2,7.7,0.0,8754.4,8754.4,0.0,0.0,0.0,11.6,28.3,39.9,33370.2
7,2026-04-01,13,W,0.0,768.3,768.3,0.0,82.4,82.4,0.0,436.4,436.4,0.0,345.6,345.6,0.0,0.0,0.0,0.0,364.5,364.5,0.0,0.0,0.0,0.0,543.9,543.9,0.0,0.0,0.0,0.0,4.8,4.8,2677.1
8,2026-04-01,14,E,3221.3,3430.8,6652.1,146.0,1.5,147.5,6.4,1280.9,1287.3,10148.7,1147.7,11296.4,0.0,2022.8,2022.8,0.0,825.8,825.8,1.5,6.2,7.8,0.0,10749.1,10749.1,0.0,0.0,0.0,25.4,27.1,52.6,33584.4
9,2026-04-01,14,W,0.0,826.8,826.8,0.0,71.3,71.3,0.0,456.1,456.1,0.0,411.0,411.0,0.0,0.0,0.0,0.0,387.0,387.0,0.0,0.0,0.0,0.0,526.6,526.6,0.0,0.0,0.0,0.0,4.7,4.7,2654.2


---
## TIE FLOWS

### 26. TieFlows (spp_physical.TieFlows)

In [18]:
df_tieflow = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_physical.TieFlows "
    "WHERE DATE(GMTTime) = '2026-04-01' ORDER BY MarketTime LIMIT 100")
print(df_tieflow.shape)
print('Columns:', df_tieflow.columns.tolist())
df_tieflow.head(10)

(100, 30)
Columns: ['MarketTime', 'GMTTime', 'SPPNSI', 'DPC', 'MCWEST', 'CLEC', 'SPC', 'AMRN', 'OTP', 'GRE', 'MDU', 'SOUC', 'SPA', 'SGE', 'AECI', 'ERCOTN', 'WAUE', 'EDDY', 'ERCOTE', 'BLKW', 'LAMAR', 'MEC', 'SPPNAI', 'RCEAST', 'EES', 'NSP', 'SCSE', 'ALTW', 'TVA', 'SIKE']


,MarketTime,GMTTime,SPPNSI,DPC,MCWEST,CLEC,SPC,AMRN,OTP,GRE,...,LAMAR,MEC,SPPNAI,RCEAST,EES,NSP,SCSE,ALTW,TVA,SIKE
0,2026-03-31 19:00:00,2026-04-01 00:00:00,398.6,31.5,20.6,80.9,-34.5,287.1,347.1,229.3,...,0.6,-934.2,389.4,102.7,253.6,-263.3,0.0,-123.4,-49.9,0.0
1,2026-03-31 19:01:00,2026-04-01 00:01:00,410.1,31.5,20.7,80.1,-27.5,284.2,350.1,229.9,...,0.4,-962.2,356.6,108.8,244.7,-265.5,0.0,-123.4,-49.5,0.0
2,2026-03-31 19:02:00,2026-04-01 00:02:00,435.5,31.5,20.4,75.6,-26.1,282.4,356.6,232.0,...,0.4,-977.9,334.9,114.9,213.4,-260.2,0.0,-120.8,-49.9,0.0
3,2026-03-31 19:03:00,2026-04-01 00:03:00,446.4,32.0,20.5,76.9,-20.6,279.8,370.2,234.4,...,0.6,-961.1,365.2,121.5,196.9,-264.2,0.0,-117.2,-49.5,0.0
4,2026-03-31 19:04:00,2026-04-01 00:04:00,464.0,32.1,20.6,74.0,-13.4,275.5,378.9,236.4,...,-0.1,-941.0,348.6,127.2,138.9,-261.8,0.0,-114.0,-48.8,0.0
5,2026-03-31 19:05:00,2026-04-01 00:05:00,489.5,32.4,20.5,71.8,-13.3,271.7,381.2,237.9,...,1.3,-971.8,349.8,133.6,137.9,-259.3,0.0,-111.7,-47.2,0.0
6,2026-03-31 19:06:00,2026-04-01 00:06:00,434.5,32.7,20.3,75.9,-22.3,267.5,388.6,238.4,...,0.6,-1004.6,298.4,134.0,141.4,-268.9,0.0,-112.5,-46.1,0.0
7,2026-03-31 19:07:00,2026-04-01 00:07:00,435.4,32.5,20.5,78.2,-39.7,265.0,387.8,238.2,...,0.9,-1108.9,141.3,134.0,118.6,-273.9,0.0,-114.1,-43.4,0.0
8,2026-03-31 19:08:00,2026-04-01 00:08:00,452.7,32.6,20.6,84.9,-55.1,266.5,396.2,238.6,...,0.6,-1092.5,239.5,134.0,174.1,-272.6,0.0,-117.7,-43.0,0.0
9,2026-03-31 19:09:00,2026-04-01 00:09:00,440.2,32.5,20.6,92.3,-50.2,274.6,399.2,247.7,...,0.4,-1071.1,302.6,134.0,193.4,-264.4,0.0,-115.1,-42.2,0.0


---
## RNU RATE

### 27. RNU Rate (spp_virtual.rnu_rate)

In [15]:
df_rnu = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_virtual.rnu_rate "
    "ORDER BY dt DESC LIMIT 50")
print(df_rnu.shape)
df_rnu.head(10)

(50, 5)


,dt,Run_Type,RT_RNU_Dist_Rate,RNU_Rate_Forecast,source
0,2026-04-07,Frcst,-0.4716,-0.4716,ForecastV1
1,2026-04-06,Frcst,-0.4716,-0.4716,ForecastV1
2,2026-04-05,Frcst,-0.4716,-0.4716,ForecastV1
3,2026-04-04,Frcst,-0.4716,-0.4716,ForecastV1
4,2026-04-03,Frcst,-0.4716,-0.4716,ForecastV1
5,2026-04-02,Frcst,-0.4716,-0.4716,ForecastV1
6,2026-04-01,Frcst,-0.4716,-0.4716,ForecastV1
7,2026-03-31,Frcst,-0.4716,-0.4716,ForecastV1
8,2026-03-30,Frcst,-0.4716,-0.4716,ForecastV1
9,2026-03-29,Frcst,-0.4716,-0.4716,ForecastV1


### 28. RNU Rate Detailed (spp_virtual.rnu_rate_detailed)

In [16]:
df_rnu_det = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_virtual.rnu_rate_detailed "
    "ORDER BY dt DESC LIMIT 50")
print(df_rnu_det.shape)
df_rnu_det.head(10)

(50, 14)


,auto_id,dt,Run_Type,DA_Inadequacy,RT_Inadequacy,RT_Out_of_Merit,RT_Reg_Dep_ADJ,RT_JOA,RT_Net_Inadvertent,RT_Congestion,RT_RNU_Dist_Qty,RT_RNU_Dist_Rate,RT_RNU_Hrly_Amt,source
0,8936,2026-03-17,S7,35.39,-2434280.0,-235974.00,-48091.50,536581.0,5114.84,-2429240.0,2091530.0,-1.0431,2181670.0,ftp
1,8937,2026-03-16,S7,14.77,-166803.0,-29995.60,-11438.60,680046.0,27075.60,-139809.0,2081520.0,0.2267,-471884.0,ftp
2,8938,2026-03-15,S7,21.30,-1430000.0,-1044.56,-23408.40,772266.0,3073.29,-1427000.0,1915030.0,-0.3562,682137.0,ftp
3,8939,2026-03-14,S7,-45.05,-713455.0,-654.43,9132.07,335411.0,-6809.07,-720302.0,1781430.0,-0.2075,369636.0,ftp
4,8940,2026-03-13,S7,-16.25,378596.0,-952.92,-23912.30,205847.0,-579785.00,-201210.0,1797160.0,0.3114,-559623.0,ftp
5,8941,2026-03-12,S7,10.95,-455971.0,0.00,-26191.30,224016.0,11495.50,-444514.0,1959210.0,-0.1318,258230.0,ftp
6,8942,2026-03-11,S7,-26.16,-1970000.0,0.00,-12929.10,640875.0,-8588.67,-1978680.0,1941920.0,-0.6911,1342080.0,ftp
7,8901,2026-03-10,S7,2.74,-1023470.0,-39491.90,-19734.50,530062.0,827.45,-1022670.0,1912100.0,-0.2890,552619.0,ftp
8,8902,2026-03-09,S7,13.30,-549151.0,-12234.50,-37190.70,309633.0,19538.00,-529687.0,1800350.0,-0.1605,288942.0,ftp
9,8903,2026-03-08,S7,18.32,-635888.0,-1673.46,-2075.56,79495.2,-8633.87,-644632.0,1707120.0,-0.3281,560107.0,ftp


---
## OP CHARGES (MWP)

### 29. Op Reserve and Fees (spp_virtual.v3_op_reserve_and_fees)

In [105]:
df_opcharge = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_virtual.v3_op_reserve_and_fees "
    "ORDER BY dt DESC LIMIT 100")
print(df_opcharge.shape)
print('Columns:', df_opcharge.columns.tolist())
df_opcharge.head(20)

(100, 4)
Columns: ['dt', 'daMwpDistRate', 'rtMwpDistRate', 'source']


,dt,daMwpDistRate,rtMwpDistRate,source
0,2026-04-11,0.3993,0.9525,forecastV2
1,2026-04-10,0.2057,1.1162,forecastV2
2,2026-04-09,0.4258,1.3488,forecastV2
3,2026-04-08,0.4790,1.0496,forecastV2
4,2026-04-07,0.3690,0.9374,forecastV2
5,2026-04-06,0.2423,0.8814,forecastV2
6,2026-04-05,0.1645,0.9928,forecastV2
7,2026-04-04,0.2818,0.8553,forecastV2
8,2026-04-03,0.4005,1.0808,forecastV2
9,2026-04-02,0.3155,0.9194,forecastV2


---
## CONSTRAINTS

### 30. DA Constraint Congestion (spp_constraints.congestion_da_hourly_v3)

Raw table — downloaded by `SPP_download_DA_Constraints_v20170324.php`. Tracks DA binding constraint shadow prices (M-values) per hour.

In [8]:
df_da_cong = sql_functions.download_df_from_sql_db(
    "SELECT * "
    "FROM spp_constraints.congestion_da_hourly_v3 "
    "WHERE dt = '{}' "
    "ORDER BY mvalue DESC LIMIT 20".format(START)
)
print(df_da_cong.shape)
display(df_da_cong.head(20))

(20, 4)


,dt,hr,ConstraintNum,MValue
0,2026-04-01,9,830532,-0.0068
1,2026-04-01,2,830532,-0.0375
2,2026-04-01,4,830532,-0.0487
3,2026-04-01,6,830532,-0.0532
4,2026-04-01,3,830532,-0.0575
5,2026-04-01,5,830532,-0.1058
6,2026-04-01,10,830535,-0.1146
7,2026-04-01,1,830532,-0.1230
8,2026-04-01,3,511187,-0.1318
9,2026-04-01,8,830532,-0.1457


### 31. DA Constraint M-Values via nighthawk `Constraint.get_mvalues()`

Reads from `congestion_da_hourly_v3`. Returns M-values (shadow prices × MW) for all constraints. Use `granularity='hourly'` for hour-by-hour breakdown.

In [9]:
from nighthawk.data.network.constraint import Constraint

constraint_spp = Constraint(oops_constraint_num_df=None, market='SPP')
df_da_mval = constraint_spp.get_mvalues(START, END, type='DA', granularity='hourly')
print(df_da_mval.shape)
display(df_da_mval.sort_values('mvalue', ascending=False).head(20))

(462, 4)


,oops_constraint_num,dt,hr,mvalue
95,830532,2026-04-01,9,-0.0068
17,830532,2026-04-01,2,-0.0375
37,830532,2026-04-01,4,-0.0487
62,830532,2026-04-01,6,-0.0532
27,830532,2026-04-01,3,-0.0575
49,830532,2026-04-01,5,-0.1058
110,830535,2026-04-01,10,-0.1146
8,830532,2026-04-01,1,-0.1230
22,511187,2026-04-01,3,-0.1318
82,830532,2026-04-01,8,-0.1457


### 32. DA Constraint Family M-Values via `ConstraintFamily.get_mvalues()`

Reads `congestion_da_hourly_v3` + `ConstraintFamilyMaster_v3`. Groups constraints by monitored facility name.

In [12]:
from nighthawk.data.network.constraint_family import ConstraintFamily

cf_spp = ConstraintFamily(constraint_family_num_df=None, opexchange='SPP')
df_cf_mval = cf_spp.get_mvalues(START, END, type='DA', granularity='daily')
print(df_cf_mval.shape)
display(df_cf_mval.sort_values('mvalue', ascending=False).head(20))

(63, 3)


,constraintFamilyNum,dt,mvalue
49,1223658,2026-04-01,-0.1627
51,1223737,2026-04-01,-0.2780
60,1802563,2026-04-01,-0.7409
23,6299,2026-04-01,-1.9271
58,1722555,2026-04-01,-2.0716
38,498407,2026-04-01,-2.0750
53,1224367,2026-04-01,-3.1680
22,1037,2026-04-01,-4.9182
9,648,2026-04-01,-7.1712
55,1357050,2026-04-01,-8.2805


### 33. RT Constraint Congestion Hourly (spp_constraints.congestion_rt_hourly_v3)

Aggregated from 5-min intervals. Same structure as DA but for RTBM.

In [ ]:
df_rt_cong_hr = sql_functions.download_df_from_sql_db(
    "SELECT *"
    "FROM spp_constraints.congestion_rt_hourly_v3 "
    "WHERE dt = '{}' "
    "ORDER BY mvalue DESC LIMIT 20".format(START)
)
print(df_rt_cong_hr.shape)
display(df_rt_cong_hr.head(20))

(20, 4)


,dt,hr,ConstraintNum,MValue
0,2026-04-01,24,433608,-0.0576
1,2026-04-01,12,867609,-0.3901
2,2026-04-01,12,867791,-0.4408
3,2026-04-01,3,634003,-0.5508
4,2026-04-01,15,867609,-0.8661
5,2026-04-01,20,819280,-0.9176
6,2026-04-01,24,351341,-1.2380
7,2026-04-01,23,634003,-1.6817
8,2026-04-01,14,867609,-2.2330
9,2026-04-01,22,634003,-2.6842


### 34. RT Constraint Congestion 5-min (spp_constraints.congestion_rt_5_min_v3)

Highest-frequency constraint data — used by the contour map for real-time display.

In [61]:
df_rt_cong_5m = sql_functions.download_df_from_sql_db(
    "SELECT * "
    "FROM spp_constraints.congestion_rt_hourly_v3 "
    "WHERE dt >= '2026-04-01' "
    "ORDER BY dt "
    "LIMIT 20"
)

print(df_rt_cong_5m.shape)
display(df_rt_cong_5m.head(40))

(20, 4)


,dt,hr,ConstraintNum,MValue
0,2026-04-01,1,653411,-305.6833
1,2026-04-01,1,825657,-135.7030
2,2026-04-01,2,653411,-836.1967
3,2026-04-01,2,681335,-47.6150
4,2026-04-01,3,351341,-25.2826
5,2026-04-01,3,634003,-0.5508
6,2026-04-01,3,653411,-145.4984
7,2026-04-01,3,681335,-19.4162
8,2026-04-01,4,351341,-34.5707
9,2026-04-01,5,351341,-56.5493


### 35. RT Constraint M-Values via nighthawk `Constraint.get_mvalues(type='RT')`

In [57]:
df_rt_mval = constraint_spp.get_mvalues(START, END, type='RT', granularity='hourly')
print(df_rt_mval.shape)
display(df_rt_mval.sort_values('mvalue', ascending=False))
display(df_rt_mval[df_rt_mval['hr']==1])

(178, 4)


,oops_constraint_num,dt,hr,mvalue
168,433608,2026-04-01,24,-0.0576
42,867609,2026-04-01,12,-0.3901
43,867791,2026-04-01,12,-0.4408
5,634003,2026-04-01,3,-0.5508
69,867609,2026-04-01,15,-0.8661
...,...,...,...,...
119,653411,2026-04-01,20,-1256.0846
161,653411,2026-04-01,23,-1270.3324
133,653411,2026-04-01,21,-1330.6849
16,863216,2026-04-01,8,-1381.4667


,oops_constraint_num,dt,hr,mvalue
0,653411,2026-04-01,1,-305.6833
1,825657,2026-04-01,1,-135.7030


### 36. Constraint Rolling Summaries (7D / 30D / 365D)

Pre-computed rolling M-value aggregates for trend analysis.

In [22]:
df_7d = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_constraints.congestion_rt_7D_v3  LIMIT 10"
)
df_30d = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_constraints.congestion_rt_30D_v3  LIMIT 10"
)
df_365d = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_constraints.congestion_rt_365D_v3  LIMIT 10"
)
print('7D:', df_7d.shape)
display(df_7d)
print('30D:', df_30d.shape)
display(df_30d)
print('365D:', df_365d.shape)
display(df_365d)

7D: (0, 7)


,ConstraintNum,TotalOn,TotalOff,OnHrs,OffHrs,OnAvg,OffAvg


30D: (0, 4)


,dt,hr,ConstraintNum,MValue


365D: (0, 4)


,dt,hr,ConstraintNum,MValue


---
## LMP (SETTLEMENT LOCATION PRICES)

**Nighthawk entry point:** `Node(node_nums=[...], market='SPP').get_price()`  
**Source tables:**
- DA hourly → `spp_lmp.settlement_location_da_hourly` (`da_value`, `congestional_value`, `marginalloss_value`)
- RT hourly → `spp_lmp.settlement_location_rt_hourly` (`rt_value`, `congestional_value`, `marginalloss_value`)
- RT daily summary → `spp_lmp.sl_rt_daily_summary` (`daily`, `on_peak`, `off_peak`, `daily_congestion`, ...)
- DA daily summary → `spp_lmp.sl_da_daily_summary`

**Note:** `rt_hrly_from_5min_table` for SPP points to the same `settlement_location_rt_hourly` (no separate fallback table).  
**Reprice:** `settlement_location_da_hourly_reprice` is written weekly; 0 rows for 2026-04-01 is expected.

### 37. DA Hourly LMP — Raw (spp_lmp.settlement_location_da_hourly)

Written by `download_LMP_from_CSV_to_DB3.php` via `REPLACE INTO`.  
**Columns:** `dt, hr, node_num, da_value (LMP), congestional_value (MCC), marginalloss_value (MLC)`  
**Expected for 2026-04-01:** 1580 nodes × 24 hrs = **37,920 rows**

In [62]:
# SPP North Hub node_num=635 — use as reference node
HUB_NODE = 635

df_da_lmp = sql_functions.download_df_from_sql_db(
    "SELECT dt, hr, node_num, da_value, congestional_value, marginalloss_value "
    "FROM spp_lmp.settlement_location_da_hourly "
    "WHERE dt='2026-04-01' AND node_num=635 "
    "ORDER BY hr"
)
# Check LMP = MCC + MLC + Slack (energy component)
df_da_lmp['implied_slack'] = df_da_lmp['da_value'] - df_da_lmp['congestional_value'] - df_da_lmp['marginalloss_value']
print('DA hourly shape:', df_da_lmp.shape, '  expected (24, 6)')
print('LMP decomp check (MCC+MLC+Slack should = LMP):')
display(df_da_lmp)

DA hourly shape: (24, 7)   expected (24, 6)
LMP decomp check (MCC+MLC+Slack should = LMP):


,dt,hr,node_num,da_value,congestional_value,marginalloss_value,implied_slack
0,2026-04-01,1,635,20.5748,-2.3471,-0.1380,23.0599
1,2026-04-01,2,635,16.8729,-0.4388,-0.1371,17.4488
2,2026-04-01,3,635,16.9680,-0.5097,-0.2124,17.6901
3,2026-04-01,4,635,17.2085,-0.2966,-0.3332,17.8383
4,2026-04-01,5,635,16.7601,-0.8054,-0.3227,17.8882
5,2026-04-01,6,635,21.5390,-0.4000,-0.4260,22.3650
6,2026-04-01,7,635,30.6228,-1.7332,-0.4494,32.8054
7,2026-04-01,8,635,30.5674,-1.8568,-0.3425,32.7667
8,2026-04-01,9,635,33.4855,-3.9063,-0.5511,37.9429
9,2026-04-01,10,635,43.5319,5.9016,-0.6434,38.2737


### 38. Row count check — DA 2026-04-01 should have exactly 37,920 rows

In [63]:
df_da_cnt = sql_functions.download_df_from_sql_db(
    "SELECT COUNT(DISTINCT node_num) as nodes, COUNT(DISTINCT hr) as hrs, COUNT(*) as total "
    "FROM spp_lmp.settlement_location_da_hourly WHERE dt='2026-04-01'"
)
print('DA 2026-04-01 count:', df_da_cnt.to_string(index=False))
assert df_da_cnt['total'][0] == 37920, f"Expected 37920, got {df_da_cnt['total'][0]}"
print('PASS: 1580 nodes x 24 hrs = 37920')

DA 2026-04-01 count:  nodes  hrs  total
  1580   24  37920
PASS: 1580 nodes x 24 hrs = 37920


### 39. DA LMP Reprice (spp_lmp.settlement_location_da_hourly_reprice)

Written by `spp_dataDownload_da_lmp_reprice.py` — runs **weekly**, not daily.  
0 rows for a given date is **normal** unless that date was repriced.  
Columns: `dt, hr, node_num, Revision, Reason, da_value, marginalloss_value, congestional_value, Modified_Date`

In [65]:
df_reprice = sql_functions.download_df_from_sql_db(
    "SELECT dt, hr, node_num, Revision, Reason, da_value, Modified_Date "
    "FROM spp_lmp.settlement_location_da_hourly_reprice "
    "WHERE dt='2026-04-01' "
    "ORDER BY hr, Revision"
)
print('Reprice rows for 2026-04-01 hub node:', df_reprice.shape[0])
print('(0 rows expected — reprice runs weekly and only if SPP re-publishes the file)')
display(df_reprice)

Reprice rows for 2026-04-01 hub node: 0
(0 rows expected — reprice runs weekly and only if SPP re-publishes the file)


,dt,hr,node_num,Revision,Reason,da_value,Modified_Date


### 40. RT 5-min LMP — Raw (spp_lmp.sl_rt_5_min)

Written by `SPP_download_DART_LMP.php` (mode=rt) and gap-filled by `SPPLMPDailyCheck_5min.py`.  
**Columns:** `forecast_time (interval end datetime), dt, hr, node_num, rt_value, congestional_value, marginalloss_value`  
**Expected:** 12 intervals per hr × 24 hrs × ~1580 nodes

In [66]:
df_rt_5m = sql_functions.download_df_from_sql_db(
    "SELECT forecast_time, dt, hr, node_num, rt_value, congestional_value, marginalloss_value "
    "FROM spp_lmp.sl_rt_5_min "
    "WHERE dt='2026-04-01' AND hr=10 AND node_num=635 "
    "ORDER BY forecast_time"
)
print('RT 5-min for hub node hr=10:', df_rt_5m.shape, '  expected (12, 7)')
display(df_rt_5m)

RT 5-min for hub node hr=10: (12, 7)   expected (12, 7)


,forecast_time,dt,hr,node_num,rt_value,congestional_value,marginalloss_value
0,2026-04-01 09:05:00,2026-04-01,10,635,26.0402,-0.7913,-1.2187
1,2026-04-01 09:10:00,2026-04-01,10,635,31.9450,-0.6733,-1.4542
2,2026-04-01 09:15:00,2026-04-01,10,635,32.5124,-0.8039,-1.5158
3,2026-04-01 09:20:00,2026-04-01,10,635,38.3603,-0.8446,-1.7496
4,2026-04-01 09:25:00,2026-04-01,10,635,43.7941,-0.5941,-2.0328
5,2026-04-01 09:30:00,2026-04-01,10,635,51.5973,-1.2350,-2.2882
6,2026-04-01 09:35:00,2026-04-01,10,635,49.4960,-1.1183,-2.1351
7,2026-04-01 09:40:00,2026-04-01,10,635,66.9718,-1.6028,-2.8926
8,2026-04-01 09:45:00,2026-04-01,10,635,75.6710,-2.1242,-3.3331
9,2026-04-01 09:50:00,2026-04-01,10,635,77.8271,-1.9417,-3.6093


### 41. RT Hourly LMP — Raw (spp_lmp.settlement_location_rt_hourly)

Derived from `sl_rt_5_min` by `AVG(rt_value) GROUP BY (dt, hr, node_num)`.  
**Expected for 2026-04-01:** up to 37,920 rows — actual is **37,622** (298 gaps expected due to RTBM timing)

In [67]:
df_rt_lmp = sql_functions.download_df_from_sql_db(
    "SELECT dt, hr, node_num, rt_value, congestional_value, marginalloss_value "
    "FROM spp_lmp.settlement_location_rt_hourly "
    "WHERE dt='2026-04-01' AND node_num=635 "
    "ORDER BY hr"
)
print('RT hourly shape:', df_rt_lmp.shape, '  expected (24, 6)')
display(df_rt_lmp)

RT hourly shape: (24, 6)   expected (24, 6)


,dt,hr,node_num,rt_value,congestional_value,marginalloss_value
0,2026-04-01,1,635,11.3352,-2.3271,-0.6324
1,2026-04-01,2,635,11.2304,-2.0801,-0.5788
2,2026-04-01,3,635,18.0865,-0.3879,-0.6114
3,2026-04-01,4,635,19.3214,0.1538,-0.6777
4,2026-04-01,5,635,25.7622,0.2508,-0.8960
5,2026-04-01,6,635,36.3466,0.2727,-1.6857
6,2026-04-01,7,635,33.4603,-0.4688,-1.4933
7,2026-04-01,8,635,48.6189,-1.0068,-1.7959
8,2026-04-01,9,635,145.2658,-0.5969,-6.0424
9,2026-04-01,10,635,54.3445,-1.6053,-2.5060


### 42. Node.get_price() — nighthawk function (DA + RT hourly)

**SPP DA path:** reads `settlement_location_da_hourly` directly  
**SPP RT path:** reads `settlement_location_rt_hourly`; if end_dt > max(dt), falls back to same table  
**Column mapping:**
- `da_value` → `da_total` (LMP)
- `congestional_value` → `da_congestion` (MCC)
- `marginalloss_value` → `da_loss` (MLC)
- `da_value - congestional_value - marginalloss_value` → `da_slack`

In [86]:
from nighthawk.data.network.node import Node

node_spp = Node(node_nums=[635], market='SPP')

# DA hourly
df_node_da = node_spp.get_price('2026-04-01', '2026-04-01',
                                component=['LMP','MCC','MLC','Slack'], type=['DA'], granularity='hourly')
print('Node DA hourly:', df_node_da.shape)
display(df_node_da)

# RT hourly
df_node_rt = node_spp.get_price('2026-04-01', '2026-04-01',
                                component=['LMP','MCC','MLC','Slack'], type=['RT'], granularity='hourly')
print('Node RT hourly:', df_node_rt.shape)
display(df_node_rt)

Node DA hourly: (24, 7)


,dt,hr,node_num,da_total,da_congestion,da_loss,da_slack
0,2026-04-01,1,635,20.574800,-2.3471,-0.1380,23.059900
1,2026-04-01,2,635,16.872900,-0.4388,-0.1371,17.448799
2,2026-04-01,3,635,16.968000,-0.5097,-0.2124,17.690100
3,2026-04-01,4,635,17.208500,-0.2966,-0.3332,17.838301
4,2026-04-01,5,635,16.760099,-0.8054,-0.3227,17.888201
5,2026-04-01,6,635,21.539000,-0.4000,-0.4260,22.365000
6,2026-04-01,7,635,30.622801,-1.7332,-0.4494,32.805401
7,2026-04-01,8,635,30.567400,-1.8568,-0.3425,32.766701
8,2026-04-01,9,635,33.485500,-3.9063,-0.5511,37.942902
9,2026-04-01,10,635,43.531898,5.9016,-0.6434,38.273701


Node RT hourly: (24, 7)


,dt,hr,node_num,rt_total,rt_congestion,rt_loss,rt_slack
0,2026-04-01,1,635,11.335200,-2.3271,-0.632400,14.294700
1,2026-04-01,2,635,11.230400,-2.0801,-0.578800,13.889300
2,2026-04-01,3,635,18.086500,-0.3879,-0.611400,19.085800
3,2026-04-01,4,635,19.321400,0.1538,-0.677700,19.845301
4,2026-04-01,5,635,25.762199,0.2508,-0.896000,26.407400
5,2026-04-01,6,635,36.346600,0.2727,-1.685700,37.759602
6,2026-04-01,7,635,33.460300,-0.4688,-1.493300,35.422401
7,2026-04-01,8,635,48.618900,-1.0068,-1.795900,51.421600
8,2026-04-01,9,635,145.265793,-0.5969,-6.042400,151.905106
9,2026-04-01,10,635,54.344501,-1.6053,-2.506000,58.455799


### 43. Cross-check: Node.get_price() vs raw table — values must match

In [87]:
# DA: congestional_value in raw == da_congestion in nighthawk
merge_da = df_da_lmp[['dt','hr','congestional_value']].merge(
    df_node_da[['dt','hr','da_congestion']], on=['dt','hr'], how='outer'
)
merge_da['dt'] = merge_da['dt'].astype(str)
merge_da['diff'] = (merge_da['congestional_value'] - merge_da['da_congestion']).abs()
print('DA MCC diff (max should be ~0):', merge_da['diff'].max())

# RT: rt_value in raw == rt_total in nighthawk
merge_rt = df_rt_lmp[['dt','hr','rt_value']].merge(
    df_node_rt[['dt','hr','rt_total']], on=['dt','hr'], how='outer'
)
merge_rt['dt'] = merge_rt['dt'].astype(str)
merge_rt['diff'] = (merge_rt['rt_value'] - merge_rt['rt_total']).abs()
print('RT LMP diff (max should be ~0):', merge_rt['diff'].max())

DA MCC diff (max should be ~0): 4.2877197259372224e-07
RT LMP diff (max should be ~0): 2.0605468762369128e-05


### 44. Last-30-Day Rolling Tables (sl_da_hourly_last30days / sl_rt_hourly_last30days)

Same schema as core hourly tables. Used by nighthawk for faster recent-date queries.  
Written by OOPSify after each DA/RT download.

In [88]:
df_da_30d = sql_functions.download_df_from_sql_db(
    "SELECT dt, hr, node_num, da_value, congestional_value "
    "FROM spp_lmp.sl_da_hourly_last30days "
    "WHERE dt='2026-04-01' AND node_num=635 ORDER BY hr"
)
df_rt_30d = sql_functions.download_df_from_sql_db(
    "SELECT dt, hr, node_num, rt_value, congestional_value "
    "FROM spp_lmp.sl_rt_hourly_last30days "
    "WHERE dt='2026-04-01' AND node_num=635 ORDER BY hr"
)
print('DA last30d:', df_da_30d.shape, '  RT last30d:', df_rt_30d.shape)
display(df_da_30d)

DA last30d: (24, 5)   RT last30d: (24, 5)


,dt,hr,node_num,da_value,congestional_value
0,2026-04-01,1,635,20.5748,-2.3471
1,2026-04-01,2,635,16.8729,-0.4388
2,2026-04-01,3,635,16.9680,-0.5097
3,2026-04-01,4,635,17.2085,-0.2966
4,2026-04-01,5,635,16.7601,-0.8054
5,2026-04-01,6,635,21.5390,-0.4000
6,2026-04-01,7,635,30.6228,-1.7332
7,2026-04-01,8,635,30.5674,-1.8568
8,2026-04-01,9,635,33.4855,-3.9063
9,2026-04-01,10,635,43.5319,5.9016


### 45. Daily Summary LMP (sl_rt_daily_summary / sl_da_daily_summary)

Written by OOPSify. Used by `Node.get_price(..., granularity='daily')`.  
**Columns:** `dt, node_num, daily, on_peak, off_peak, daily_congestion, on_peak_congestion, off_peak_congestion, daily_marginloss, ...`  
**Expected for 2026-04-01:** 1,580 rows (one per node)

In [89]:
df_rt_daily = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_lmp.sl_rt_daily_summary "
    "WHERE dt='2026-04-01' AND node_num=635"
)
print('RT daily summary (node 635):', df_rt_daily.shape)
display(df_rt_daily)

# Row count check
df_rt_daily_cnt = sql_functions.download_df_from_sql_db(
    "SELECT COUNT(*) as total FROM spp_lmp.sl_rt_daily_summary WHERE dt='2026-04-01'"
)
print('Total RT daily rows:', df_rt_daily_cnt['total'][0], ' expected ~1580')

RT daily summary (node 635): (1, 11)


,dt,node_num,on_peak,off_peak,daily,on_peak_congestion,off_peak_congestion,daily_congestion,on_peak_marginloss,off_peak_marginloss,daily_marginloss
0,2026-04-01,635,1428.73,116.118,1544.84,-7.4291,-21.194,-28.6231,-73.9106,-5.5102,-79.4208


Total RT daily rows: 1580  expected ~1580


### 46. Node.get_price() daily granularity — reads sl_rt_daily_summary

In [ ]:
df_node_daily = node_spp.get_price('2026-04-01', '2026-04-01',
                                   component=['LMP','MCC','MLC'], type=['DA','RT'], granularity='daily')
print('Node daily (DA+RT):', df_node_daily.shape)
display(df_node_daily)

### 47. RT 5-min Slack Last 2 Years — settlement_location_rt_5min_slack_last2years
> Script `spp_dataDownload_rt_slack_last2years.py` is **deprecated** (2026-03-12); table is now refreshed by `spp_update_lmp_summary.py` (`update_rt_5min_slack_last2years`).

In [97]:
df_rt_slack = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_lmp.settlement_location_rt_5min_slack_last2years where dt = '2026-04-01' and hr = 1"
)

print('RT 5-min slack last-2yr:', df_rt_slack.shape)
display(df_rt_slack)

df_rt_slack_cnt = sql_functions.download_df_from_sql_db(
    "SELECT COUNT(*) as total, MIN(dt) as min_dt, MAX(dt) as max_dt "
    "FROM spp_lmp.settlement_location_rt_5min_slack_last2years"
)

display(df_rt_slack_cnt)

RT 5-min slack last-2yr: (6, 4)


,Interval,dt,hr,slack
0,2026-04-01 00:35:00,2026-04-01,1,14.1196
1,2026-04-01 00:40:00,2026-04-01,1,14.0267
2,2026-04-01 00:45:00,2026-04-01,1,18.4962
3,2026-04-01 00:50:00,2026-04-01,1,11.6634
4,2026-04-01 00:55:00,2026-04-01,1,13.5899
5,2026-04-01 01:00:00,2026-04-01,1,13.8722


,total,min_dt,max_dt
0,209234,2024-04-06,2026-04-06


### 48. Multi-Day LMP Forecast — settlement_location_multi_day_forecast

In [98]:
df_mdfc = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_lmp.settlement_location_multi_day_forecast "
    "ORDER BY as_of_dt DESC, dt, hr LIMIT 5"
)
print('Multi-day forecast (all):', df_mdfc.shape)
display(df_mdfc)

df_mdfc_virt = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_lmp.settlement_location_multi_day_forecast_virtual "
    "ORDER BY dt DESC, hr LIMIT 5"
)
print('Multi-day forecast (virtual):', df_mdfc_virt.shape)
display(df_mdfc_virt)

Multi-day forecast (all): (5, 8)


,as_of_dt,dt,hr,node_num,LMP,MLC,MCC,MEC
0,2026-04-04 18:31:39,2026-04-05,1,993,19.8918,0.6763,0.0814,19.1341
1,2026-04-04 18:31:39,2026-04-05,1,2,20.1957,0.9712,0.0904,19.1341
2,2026-04-04 18:31:39,2026-04-05,1,7,19.6284,0.4126,0.0816,19.1342
3,2026-04-04 18:31:39,2026-04-05,1,6,20.4058,1.1749,0.0967,19.1342
4,2026-04-04 18:31:39,2026-04-05,1,1,20.1589,0.8906,0.1342,19.1341


Multi-day forecast (virtual): (5, 7)


,dt,hr,node_num,LMP,MLC,MCC,MEC
0,2026-04-07,1,2,14.6243,0.6407,2.9938,10.9898
1,2026-04-07,1,8,20.4103,0.7280,8.6926,10.9897
2,2026-04-07,1,7,12.6757,0.3130,1.3730,10.9897
3,2026-04-07,1,1,12.8881,0.5707,1.3277,10.9897
4,2026-04-07,1,6,13.2877,0.7454,1.5526,10.9897


### 49. DA MCP Zonal Hourly — reserve_mcp_da_zonal_hourly

In [99]:
df_da_mcp = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_lmp.reserve_mcp_da_zonal_hourly "
    "ORDER BY dt DESC, hr, Reserve_Zone LIMIT 5"
)
print('DA MCP zonal hourly:', df_da_mcp.shape)
display(df_da_mcp)

df_da_mcp_cnt = sql_functions.download_df_from_sql_db(
    "SELECT COUNT(*) as total, MIN(dt) as min_dt, MAX(dt) as max_dt "
    "FROM spp_lmp.reserve_mcp_da_zonal_hourly"
)
display(df_da_mcp_cnt)

DA MCP zonal hourly: (5, 10)


,dt,hr,Reserve_Zone,RegUP,RegDN,RampUP,RampDN,Spin,Supp,UncUP
0,2026-04-06,1,1,5.2983,3.8823,0.0,0.0,2.0001,0.449,0.0
1,2026-04-06,1,2,5.2983,3.8823,0.0,0.0,2.0001,0.449,0.0
2,2026-04-06,1,3,5.2983,3.8823,0.0,0.0,2.0001,0.449,0.0
3,2026-04-06,1,4,5.2983,3.8823,0.0,0.0,2.0001,0.449,0.0
4,2026-04-06,1,5,5.2983,3.8823,0.0,0.0,2.0001,0.449,0.0


,total,min_dt,max_dt
0,371117,2013-07-01,2026-04-06


### 50. DA Market Clearing — spp_virtual.market_clearing

In [100]:
df_mc = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_virtual.market_clearing "
    "ORDER BY dt DESC, hr LIMIT 5"
)
print('DA market clearing:', df_mc.shape)
display(df_mc)

df_mc_cnt = sql_functions.download_df_from_sql_db(
    "SELECT COUNT(*) as total, MIN(dt) as min_dt, MAX(dt) as max_dt "
    "FROM spp_virtual.market_clearing"
)
display(df_mc_cnt)

DA market clearing: (5, 25)


,dt,hr,baa_zone,Generation,Cleared_DR,Cleared_Demand_Bid,Cleared_Fixed_Demand_Bid,Cleared_Virtual_Bid,Cleared_Virtual_Offer,Total_Demand,NSI,SMP,Min_LMP,Max_LMP,RegUP,RegDN,RampUP,RampDN,Spin,Supp,UncUP,Capacity_Available,Curtailed_Fixed_Demand_Bid,Net_Capacity,Fixed_Obligation
0,2026-04-06,1,E,24068.80,0.0,32.0,25724.70,2060.99,5131.420,27817.7,597.975,10.879,-122.398,248.430,360.0,473.0,584.0,852.0,713.0,713.0,273.0,50646.0,0,24339,26307
1,2026-04-06,1,W,3017.33,0.0,0.0,2546.22,219.70,14.421,2765.9,257.337,23.020,18.272,24.667,76.0,86.0,40.0,147.0,87.0,87.0,0.0,4172.0,0,1626,2546
2,2026-04-06,2,E,23278.80,0.0,32.0,25363.60,1834.27,5319.290,27229.9,543.764,9.143,-101.533,202.729,359.0,391.0,563.0,652.0,713.0,713.0,1214.0,50660.0,0,24772,25888
3,2026-04-06,2,W,2931.48,0.0,0.0,2472.30,245.70,8.421,2718.0,226.220,21.676,19.475,23.161,72.0,78.0,54.0,154.0,82.0,82.0,20.0,4081.0,0,1609,2472
4,2026-04-06,3,E,23925.40,0.0,32.0,25281.10,2187.14,4884.670,27500.2,544.110,10.960,-94.718,197.412,388.0,357.0,620.0,634.0,713.0,713.0,1119.0,50261.0,0,24519,25742


,total,min_dt,max_dt
0,55049,2020-01-01,2026-04-06


### 51. DA Virtual Clearing by MOA — spp_virtual.moa_virtual_clearing

In [ ]:
df_moa = sql_functions.download_df_from_sql_db(
    "SELECT * FROM spp_virtual.moa_virtual_clearing "
    "ORDER BY dt DESC, hr, moa LIMIT 5"
)
print('MOA virtual clearing:', df_moa.shape)
display(df_moa)

df_moa_cnt = sql_functions.download_df_from_sql_db(
    "SELECT COUNT(*) as total, MIN(dt) as min_dt, MAX(dt) as max_dt "
    "FROM spp_virtual.moa_virtual_clearing"
)
display(df_moa_cnt)

MOA virtual clearing: (5, 6)


,dt,hr,BAA,moa,cleared_virtual_bid,cleared_virtual_offer
0,2026-04-06,1,E,CSWS,146.481,200.567
1,2026-04-06,1,E,EDE,67.887,104.628
2,2026-04-06,1,E,GRDA,32.114,13.225
3,2026-04-06,1,E,INDN,10.400,54.200
4,2026-04-06,1,E,KACY,25.048,5.450


,total,min_dt,max_dt
0,1600301,2015-07-03,2026-04-06


### 52. Hourly Available Outage Maintenance Margin — `get_data_and_mapping_for_avail_maintenance_margin()`
> Source tables: `spp_physical.hourly_available_outage_maintenance_margin_virtual` (forecast) and `hourly_available_outage_maintenance_margin_latest` (actual). SWPW tables not used in nighthawk.

In [102]:
from nighthawk.data.pipeline.var_handler.avail_maint_margin import (
    get_data_and_mapping_for_avail_maintenance_margin
)

data_amm, mapping_amm = get_data_and_mapping_for_avail_maintenance_margin(
    node_list=[635],
    opexchange='SPP',
    start_dt=START,
    end_dt=END,
    var_spec=['f', 'a'],
)
print('data shape:', data_amm.shape)
print('columns:', data_amm.columns.tolist())
display(data_amm)

print('\nmapping:')
display(mapping_amm)

data shape: (24, 4)
columns: ['dt', 'hr', 'spp_total_avail_maintenance_margin_f', 'spp_total_avail_maintenance_margin_a']


,dt,hr,spp_total_avail_maintenance_margin_f,spp_total_avail_maintenance_margin_a
0,2026-04-01,1,13031.0,9221.0
1,2026-04-01,2,13987.0,9706.0
2,2026-04-01,3,14088.0,8781.0
3,2026-04-01,4,13854.0,8706.0
4,2026-04-01,5,15135.0,8548.0
5,2026-04-01,6,14792.0,7217.0
6,2026-04-01,7,13106.0,4150.0
7,2026-04-01,8,11829.0,2068.0
8,2026-04-01,9,10983.0,1248.0
9,2026-04-01,10,10251.0,1411.0



mapping:


,node_num,var_name
0,635,spp_total_avail_maintenance_margin_f
1,635,spp_total_avail_maintenance_margin_a
